<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="evoTS">
<title>evoTS: analysis of evolutionary sequences of phenotypic change • evoTS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="evoTS: analysis of evolutionary sequences of phenotypic change">
<meta property="og:description" content="evoTS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-inverse navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">evoTS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../articles/evoTS_vignette.html">
    <span class="fa fa-solid fa-shoe-prints"></span>
     
    Walk-through
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">
    <span class="fa fa-regular fa-newspaper"></span>
     
    News
  </a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/klvoje/evoTS">
    <span class="fa fa-github fa-lg"></span>
     
    Github
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>evoTS: analysis of evolutionary sequences of phenotypic change</h1>
                        <h4 data-toc-skip class="author">Kjetil L.
Voje</h4>
            
            <h4 data-toc-skip class="date">2022-02-17</h4>
      
      
      <div class="d-none name"><code>evoTS_vignette.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="about-evots">1.0 About <code>evoTS</code><a class="anchor" aria-label="anchor" href="#about-evots"></a>
</h2>
<p>The <code>evoTS</code> package facilitates univariate and
multivariate analysis of evolutionary sequences of phenotypic change.
The goal of the <code>evoTS</code> package is to offer a large range of
evolutionary models to enable detailed studies of evolutionary change
within lineages, in both the fossil record and in modern lineages.</p>
<p>The <code>evoTS</code> package extends the univariate modeling
framework implemented in the
<a href="https://cran.r-project.org/web/packages/paleoTS/index.html" class="external-link"><code>paleoTS</code>
package</a> (Hunt 2006; 2008a; 2008b; Hunt <em>et al.</em> 2008; 2010;
2015) by containing functions to fit a large range of univariate
evolutionary models not implemented in <code>paleoTS</code>.
<code>evoTS</code> has been developed to mirror the user experience from
<code>paleoTS</code> as much as possible. For example, all univariate
models implemented in <code>evoTS</code> can be fitted to a
<code>paleoTS</code> object, i.e. the data format used in
<code>paleoTS</code>. The fit of all univariate models available in
<code>paleoTS</code> and <code>evoTS</code> are directly comparable.</p>
<p><code>evoTS</code> contains a range of multivariate models, including
different versions of multivariate unbiased random walks and
Ornstein-Uhlenbeck processes. Together, these models allow the user to
test various hypotheses of trait evolution, e.g. whether traits change
in a correlated or uncorrelated manner, whether one trait/variable
affects the optimum of a second trait (Granger causality), whether
adaptation in different traits happen independently toward fixed optima
etc.</p>
<p><code>evoTS</code> also contains functions for calculating the
topology of the likelihood surfaces of fitted models, a useful feature
to investigate the range of parameter values with approximately equal
likelihood as the best parameter estimates. <br><br></p>
<div class="section level3">
<h3 id="compatibility-with-paleots-technical-info">1.1 Compatibility with <code>paleoTS</code> <strong>(technical
info)</strong><a class="anchor" aria-label="anchor" href="#compatibility-with-paleots-technical-info"></a>
</h3>
<p>As in the <code>paleoTS</code> package, all models implemented in
<code>evoTS</code> assume that the population (sample) means in the
sequence of ancestor-descendants have a joint distribution that is
multivariate normal with an expected mean vector and covariance matrix
that are functions of the parameters of each model, the time intervals
separating the populations (samples) in the sequence, and the sampling
variances of the trait means calculated for each population (sample).
Given the assumption of multivariate normality, the expected
distribution of sample means is given by their first, second, and mixed
moments (covariance). As in the <code>paleoTS</code> package,
<code>evoTS</code> use the built-in optimization routines in R for
estimating maximum likelihood parameter estimates. The default
hill-climbing optimization technique used in all univariate models in
<code>evoTS</code> (L-BFGS-B) is a quasi-Newton method that constrains
the optimization of certain parameters (e.g. so that variance parameters
cannot be smaller than 0). All multivariate models use the Nelder-Mead
hill climbing search algorithm as default.</p>
<p>All models in <code>evoTS</code> have been implemented using the
joint parameterization routine from the <code>paleoTS</code> package.
The optimization is therefore fit using the actual sample values, with
the autocorrelation among samples accounted for in the log-likelihood
function. The joint distribution of sample means is multivariate normal,
with means and variance-covariances determined by evolutionary
parameters and sampling errors.</p>
<p>As in the <code>paleoTS</code> package, relative model fit in
<code>evoTS</code> are evaluated based on the small sample-corrected
version of the AICc (Akaike 1974; Burnham and Anderson 2002).</p>
<p><br><br></p>
</div>
</div>
<div class="section level2">
<h2 id="installation">2.0 Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The <code>evoTS</code> package is available on GitHub and can be
installed using devtools: <br></p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">install.packages</span>(<span class="st">"devtools"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"klvoje/evoTS"</span>)</span></code></pre></div>
<p><br> The package will soon be available on CRAN. <br><br></p>
</div>
<div class="section level2">
<h2 id="getting-data-into-evots">3.0 Getting data into <code>evoTS</code><a class="anchor" aria-label="anchor" href="#getting-data-into-evots"></a>
</h2>
<p>An object of class <code>paleoTS</code> is the required input for
most of the functions in <code>evoTS</code>. To create a
<code>paleoTS</code> object, you need vectors of trait means,
sample/population variances, sample sizes and ages of the
samples/populations.</p>
<p>One easy way to create a <code>paleoTS</code> object is to use the
<code>as.paleoTS</code> function from the <code>paleoTS</code> package.
<br></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Creating example data </span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> trait_means<span class="ot">&lt;-</span><span class="fu">rnorm</span>(<span class="dv">20</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> trait_variance<span class="ot">&lt;-</span><span class="fu">rep</span>(<span class="fl">0.5</span>,<span class="dv">20</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> sample_size<span class="ot">&lt;-</span><span class="fu">rep</span>(<span class="dv">30</span>,<span class="dv">20</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> time_vector<span class="ot">&lt;-</span><span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">19</span>,<span class="dv">1</span>)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Create paleoTS object</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> indata.evoTS<span class="ot">&lt;-</span>paleoTS<span class="sc">::</span><span class="fu">as.paleoTS</span>(<span class="at">mm =</span> trait_means, <span class="at">vv =</span> trait_variance, <span class="at">nn =</span> sample_size, <span class="at">tt =</span> time_vector)</span></code></pre></div>
<p><br> Another way to get data into <code>evoTS</code> is to use the
function <code>read.paleoTS</code> from the <code>paleoTS</code>
package. This function imports data from a text file with four columns
corresponding to trait means, sample/population variances, sample sizes
and ages of the samples/populations, and converts the input to a
<code>paleoTS</code> object.</p>
<p>See also the <code>paleoTS</code> (<a href="https://cran.r-project.org/web/packages/paleoTS/vignettes/paleoTS_basics.html" class="external-link">vignette</a>)
for more info on how to import data and create a <code>paleoTS</code>
object.</p>
<div class="section level3">
<h3 id="data-included-in-evots">3.1 Data included in <code>evoTS</code><a class="anchor" aria-label="anchor" href="#data-included-in-evots"></a>
</h3>
<p>Two evolutionary sequences (time-series) of phenotypic change are
included in the <code>evoTS</code> package. The data are from the diatom
lineage <em>Stephanodiscus yellowstonensis</em> and were originally
published in Theriot <em>et al.</em> (2006). Each trait consists of 63
samples spanning almost 14 000 years of phenotypic change.</p>
<p>We will use these data to illustrate many of the functions
implemented in <code>evoTS.</code></p>
<p>We first investigate phenotypic change in the diameter of <em>S.
yellowstonensis</em>. The diameter has been measured in micrometers, but
we are interested in investigating proportional changes in the trait. We
therefore first do an approximate log-transformation of the data. We
then convert the time vector in the data set to unit length (i.e. the
length in time from the oldest to youngest sample/population in the data
set becomes 1). Such a linear transformation of the time vector does not
change how the estimated parameters describe the evolutionary dynamics
in the data, but ease parameter estimation and interpretation of certain
model parameters. We also plot the data to have a look at the phenotypic
changes. <br></p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Doing an approximate log-transformation of the data</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> ln.diameter<span class="ot">&lt;-</span>paleoTS<span class="sc">::</span><span class="fu">ln.paleoTS</span>(diameter_S.yellowstonensis)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Convert the time vector to unit length</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> ln.diameter<span class="sc">$</span>tt<span class="ot">&lt;-</span>ln.diameter<span class="sc">$</span>tt<span class="sc">/</span>(<span class="fu">max</span>(ln.diameter<span class="sc">$</span>tt))</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Plotting the data</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">plotpaleoTS</span>(ln.diameter)</span></code></pre></div>
<p><br><img src="diameter.png" width="70%" style="display: block; margin: auto;"><br></p>
</div>
</div>
<div class="section level2">
<h2 id="univariate-models-in-evots">4.0 Univariate models in <code>evoTS</code><a class="anchor" aria-label="anchor" href="#univariate-models-in-evots"></a>
</h2>
<p>The <code>evoTS</code> package contains a range of univariate models
that supplement the models available in <code>paleoTS.</code></p>
<p>The <code>paleoTS</code> package contains functions to fit biased
(GRW) and unbiased random walks (URW), stasis (modeled as a white noise
process, i.e. uncorrelated variation around a constant mean), strict
stasis (no real evolutionary change) and an Ornstein-Uhlenbeck (OU)
processes assuming a fixed optimum (see Hunt 2006 and Hunt <em>et
al.</em> 2008 for info on these models). The <code>paleoTS</code>
package also contains models of a punctuated mode of evolution where
punctuations (jumps in phenotype space) separate different
parameterizations of the stasis model (Hunt 2008). The
<code>paleoTS</code> package also contains a few other mode shift
models, i.e. where the model of evolution shifts at some point during
the evolutionary sequence (Hunt 2008; Hunt <em>et al.</em> 2015).</p>
<p>Several new univariate models have been implemented in the
<code>evoTS</code> package. Functions are also available to fit mode
shifts between all univariate models available in <code>evoTS</code> and
<code>paleoTS</code> combined. There are also functions available in
<code>evoTS</code> that allow for more than one mode shift during the
analyzed evolutionary sequence.</p>
<p>The following univariate models have been implemented in the
<code>evoTS</code> package:</p>
<ol style="list-style-type: decimal">
<li><p>A decerelated-evolution model (an unbiased random walk with an
exponential decrease in the rate of change over time)</p></li>
<li><p>A accelerated-evolution model (an unbiased random walk with an
exponential increase in the rate of change over time)</p></li>
<li><p>The Bokma model (a punctuated equilibrium model where two (or
three) unbiased random walks are separated by one (or two)
jump(s)/punctuation(s) in morphospace)</p></li>
<li><p>Ornstein-Uhlenbeck processes where the optimum changes according
to an unbiased random walk.</p></li>
</ol>
<div class="section level3">
<h3 id="decelerated-evolution-model">4.1 Decelerated-evolution model<a class="anchor" aria-label="anchor" href="#decelerated-evolution-model"></a>
</h3>
<p>The decelerated model of phyletic evolution is an unbiased random
walk where the step variance (i.e. the variance of the normal
distribution from which evolutionary steps are drawn) is reduced
exponentially through time (Voje 2020). This model is closely related to
the early burst model developed for phylogenetic comparative data
(e.g. Harmon <em>et al.</em> 2010, Cooper and Purvis 2010), and
describes a lineage that shows a reduced rate of evolution with time. As
for the random walk model (Hunt 2006), the expected evolutionary
divergence between ancestor and descendant populations is always zero in
the model of decelerated evolution. The expected trait mean and its
variance and covariance are given by the following expressions: <br><br></p>
<p><span class="math display">\[ E[z_{i}] = z_{0} \]</span> <br></p>
<p><span class="math display">\[ Var[z_{i}] = \sigma ^{2}
_{step.0}e^{rt_i} + \epsilon _{i} \]</span> <br></p>
<p><span class="math display">\[ Cov[z_{i},z_{j}] = \sigma ^{2}
_{step.0}e^{rt_{min}} \]</span> <br></p>
<p>where <span class="math inline">\(z_{i}\)</span> is the expected
trait value for population <em>i</em> in the time series, <span class="math inline">\(z_{i}\)</span> is the ancestral trait mean, <span class="math inline">\(\sigma ^{2} _{step.0}\)</span> is the initial
value for the step distribution, <span class="math inline">\(r\)</span>
describes the exponential decay in the <span class="math inline">\(\sigma ^{2}_{step.0}\)</span> through time and is
constrained to be 0 or smaller, <span class="math inline">\(t_{i}\)</span> is the time interval from the
ancestral population mean (the start of the fossil sequence, which has a
time of 0) to the <em>i</em>th population, and <span class="math inline">\(t_{min}\)</span> is the time interval from the
ancestral population to the oldest of the two populations <span class="math inline">\(z_{i}\)</span> and <span class="math inline">\(z_{j}\)</span>.</p>
<p>The decerelated model of evolution can be fitted to data using the
<code>opt.joint.decel</code> function. <br></p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">opt.joint.decel</span>(ln.diameter)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>logL</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">76.77642</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>AICc</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">147.1461</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>parameters</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>       anc      vstep          r </span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a> <span class="fl">3.8868739</span>  <span class="fl">0.2892461</span> <span class="sc">-</span><span class="fl">1.8076302</span> </span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>modelName</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Decel"</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>method</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Joint"</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>K</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">3</span></span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>n</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">63</span></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(,<span class="st">"class"</span>)</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"paleoTSfit"</span></span></code></pre></div>
<p><br> The output returns the log-likelihood of the model parameter
(<em>logl</em>), the AICc score (<em>AICc</em>), the number (<em>K</em>)
of estimated parameters (<em>parameters</em>), the length of the
analysed time-series (<em>n</em>), the model name (<em>modelName</em>)
and the method used to parameterize the model (<em>method</em>). All
<em>anc</em> is the estimated ancestral trait value, <em>vstep</em> is
the initial value for the step distribution, and <em>r</em> describes
the exponential decay in the <em>vstep</em> parameter through time.</p>
<p>The time it takes to half the rate of evolution can be calculated
based on the value of <em>r</em> using <span class="math inline">\(-ln(2)/r\)</span>. The half-life parameter is
interpreted based on the time-scale used when analyzing the data. Since
time from start to end in our data has been scaled to unit length, the
estimated half-life represent the percent of the total length of the
time-series it takes for the rate of evolution to half. The half-life is
0.38 in this example. The total length of the analyzed time-series is
13,728 years, which means it takes 5,217 years for the rate of evolution
to be halved.</p>
<p>What are the uncertainty of the estimated parameters in this model?
One way to assess that is to set the <code>hess</code> argument to
<code>TRUE</code> when running the model
(<code>opt.joint.decel(ln.diameter, hess = TRUE)</code>). The Hessian
matrix is a square matrix of partial second order derivative and it is
possible to estimate the standard error of the model parameters by
taking the square-root of the inverse of the matrix diagonal. Another
way to assess the uncertainty of the estimated parameters is to explore
the likelihood-surface of the fitted model, which is a possibility in
<code>evoTS</code>.</p>
<p>Investigating the likelihood surface can be helpful for several
reasons.</p>
<ol style="list-style-type: decimal">
<li><p>Computing the likelihood surface is a great way to explore which
parameter combinations that have an almost identical likelihood compared
to the maximum likelihood values. Investigating the log-likelihood
surface is also an approach to assess uncertainty in the estimated
parameters. A large range of parameter values that have almost the same
log-likelihood is an indication that we should be careful when
interpreting the maximum-likelihood (best) estimates of the parameters.
The functions in <code>evoTS</code> calculating log-likelihood surfaces
report the upper and lower parameter estimates that are within two
support units of the best estimate as a way to assess uncertainty in
parameters (Edwards 1992). While standard errors computed from the
Hessian matrix are always symmetric around the point estimate, the
log-likelihood surface might not be (multivariate) normal. The reported
upper and lower parameter estimates are therefore often not symmetrical
around the maximum likelihood parameter estimates.</p></li>
<li><p>Estimating parameters in a model using maximum likelihood always
run the risk of returning parameters from a local and not a global
optimum. Investigating the support surface for combinations of
parameters is one way to explore the topology of the
likelihood-surface.</p></li>
<li><p>Ridges in the log-likelihood surface can make it challenging to
identify maximum likelihood estimates of the model parameters in certain
cases. Flat ridges may for example cause identifiability issues and
problems for the model to converge. Investigating the log-likelihood
surface can therefore help diagnose challenges related to failures of
models convergence.</p></li>
</ol>
<p>The <code>evoTS</code> package contain functions to create likelihood
surfaces for univariate models in <code>evoTS</code> and
<code>paleoTS</code> (e.g. <code>confidence.stasis</code>,
<code>confidence.URW</code>, <code>confidence.GRW</code>,
<code>confidence.OU</code>). These functions need a <code>paleoTS</code>
object and a set of candidate values for the parameters to be evaluated.
One way to define the candidate values is to use the <code>seq</code>
function. The vectors given to the arguments <code>vstep.vec</code> and
<code>r.vec</code> defines the pairwise combinations of parameters for
which the function will estimate the log-likelihood. The resolution of
the input vectors therefore determines how accurate the visual
representation of the support surface is, including the returned upper
and lower estimates printed in the console. A higher resolution gives
better precision, but demands more computation time. Note that the
computed support surface is conditional on the best estimates of the
other model parameters that are not part of the support surface
(e.g. the estimated ancestral trait value is assumed to be 3.8868739 in
the example below). <br></p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">loglik.surface.decel</span>(ln.diameter, <span class="at">vstep.vec =</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="fl">1.2</span>,<span class="fl">0.01</span>), <span class="at">r.vec =</span> <span class="fu">seq</span>(<span class="sc">-</span><span class="dv">5</span>,<span class="dv">0</span>,<span class="fl">0.01</span>))</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>      lower upper</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>vstep  <span class="fl">0.17</span>  <span class="fl">1.00</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>r     <span class="sc">-</span><span class="fl">3.70</span> <span class="sc">-</span><span class="fl">0.28</span></span></code></pre></div>
<p><br><img src="decel_logl_surface.png" width="80%" style="display: block; margin: auto;"><br> From the likelihood surface and from the printed confidence
intervals, we see that <em>r</em> values between -3.70 and -0.28 are
within 2 log-likelihood units from the best estimate for this parameter.
This suggests it is hard to exclude the possibility that the half-life
of the decay in the rate of evolution is as much as 248% (34,045 years)
or as low as 19% (2,608 years) of the time-interval investigated.</p>
</div>
<div class="section level3">
<h3 id="accelerated-evolution-model">4.2 Accelerated-evolution model<a class="anchor" aria-label="anchor" href="#accelerated-evolution-model"></a>
</h3>
<p>The accelerated evolution model is identical to the decelerated model
except that the r parameter is constrained to be 0 or larger, which
means the rate of evolution is accelerating with time.</p>
<p>The accelerated evolution model can be fitted using the
<code>opt.joint.accel</code> function. <br></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">opt.joint.accel</span>(ln.diameter)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>logL</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">73.80287</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>AICc</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">141.199</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>parameters</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>      anc     vstep         r </span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="fl">3.7078973</span> <span class="fl">0.7903365</span> <span class="fl">0.2630122</span> </span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>modelName</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Accel"</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>method</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Joint"</span></span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>K</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">3</span></span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>n</span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">63</span></span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(,<span class="st">"class"</span>)</span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"paleoTSfit"</span></span></code></pre></div>
<p><br> The accelerated evolution model has a lower (worse)
log-likelihood and higher (worse) AICc score compared to the decelerated
model of evolution.</p>
<p>A support surface can be produced using the
<code>confidence.accel</code> function. <br></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">loglik.surface.accel</span>(ln.diameter, <span class="at">vstep =</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">5</span>,<span class="fl">0.01</span>), <span class="at">r.vec =</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="fl">1.5</span>, <span class="fl">0.005</span>))</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>      lower upper</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>vstep <span class="fl">0.090</span>  <span class="fl">5.00</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>r     <span class="fl">0.035</span>  <span class="fl">1.35</span></span></code></pre></div>
<p><br><img src="accel_logl_surface.png" width="80%" style="display: block; margin: auto;"><br> The 3D plot can be rotated vertically and horizontally to get a
better overview of the likelihood surface, which is why the observation
angle is different for this 3D plot compared to the 3D plot for the
decelerated model.</p>
</div>
<div class="section level3">
<h3 id="the-bokma-model">4.3 The Bokma model<a class="anchor" aria-label="anchor" href="#the-bokma-model"></a>
</h3>
<p>The <code>paleoTS</code> package contains models that portray a
punctuated mode of evolution as a white noise process (stasis)
punctuated by jumps in phenotype space (the function
<code>opt.joint.punc</code>). Bokma (2002) suggested a different model
as a punctuated mode of evolution along branches on a phylogeny (see
also Ingram 2011 and Hunt 2013). According to Bokma’s (2002) model, a
trait is evolving as a unbiased random walk in between punctuation. This
model is implemented in the <code>evoTS</code> package and can be ran
using the function <code>fit.Bokma</code>. The function investigates and
compares all possible locations for the punctuation and returns the
shift point and the maximum likelihood parameter values for the two
unbiased random walks. The minimum length of the time-series (i.e. the
number of samples/populations) that is used to parameterize each
unbiased random walk is defined by the argument <code>minb.</code> The
user can also define which sample/population in the time series that
represents the punctuation event using the argument
<code>shift.point</code> (e.g. <code>shift.point = 20</code>). Fitting
Bokma’s model with two punctuations and three unbiased random walks can
be done using the function <code>fit.2.Bokma</code>. <br></p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> Bokma.PE<span class="ot">&lt;-</span><span class="fu">fit.Bokma</span>(ln.diameter, <span class="at">minb=</span><span class="dv">10</span>)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> Bokma.PE</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Searching for all possible switchpoints in time series"</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>Total <span class="co"># hypotheses:  44 </span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>  <span class="dv">2</span>  <span class="dv">3</span>  <span class="dv">4</span>  <span class="dv">5</span>  <span class="dv">6</span>  <span class="dv">7</span>  <span class="dv">8</span>  <span class="dv">9</span>  <span class="dv">10</span>  <span class="dv">11</span>  <span class="dv">12</span>  <span class="dv">13</span>  <span class="dv">14</span>  <span class="dv">15</span>  <span class="dv">16</span>  <span class="dv">17</span>  <span class="dv">18</span>  <span class="dv">19</span>  <span class="dv">20</span>  <span class="dv">21</span>  <span class="dv">22</span>  <span class="dv">23</span>  <span class="dv">24</span>  <span class="dv">25</span>  <span class="dv">26</span>  <span class="dv">27</span>  <span class="dv">28</span>  <span class="dv">29</span>  <span class="dv">30</span>  <span class="dv">31</span>  <span class="dv">32</span>  <span class="dv">33</span>  <span class="dv">34</span>  <span class="dv">35</span>  <span class="dv">36</span>  <span class="dv">37</span>  <span class="dv">38</span>  <span class="dv">39</span>  <span class="dv">40</span>  <span class="dv">41</span>  <span class="dv">42</span>  <span class="dv">43</span>  <span class="dv">44</span>  </span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>logL</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">83.38311</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>AICc</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">155.7136</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>parameters</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>       anc    vstep_1    new.opt    vstep_2     shift1 </span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a> <span class="fl">3.7157502</span>  <span class="fl">0.3419322</span>  <span class="fl">4.1349082</span>  <span class="fl">0.1668632</span> <span class="fl">18.0000000</span> </span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>modelName</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"PE.Bokma"</span></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>method</span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Joint"</span></span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>K</span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">5</span></span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>n</span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">63</span></span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>all.logl</span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a> [<span class="dv">1</span>] <span class="fl">80.31088</span> <span class="fl">79.51765</span> <span class="fl">80.01722</span> <span class="fl">79.66074</span> <span class="fl">80.48083</span> <span class="fl">79.89100</span> <span class="fl">79.42840</span> <span class="fl">83.38311</span> <span class="fl">80.77101</span> <span class="fl">80.77826</span> <span class="fl">79.85728</span> <span class="fl">79.68613</span> <span class="fl">79.59643</span></span>
<span id="cb8-31"><a href="#cb8-31" aria-hidden="true" tabindex="-1"></a>[<span class="dv">14</span>] <span class="fl">79.44200</span> <span class="fl">82.63356</span> <span class="fl">81.44387</span> <span class="fl">81.21899</span> <span class="fl">81.02157</span> <span class="fl">82.02478</span> <span class="fl">80.99060</span> <span class="fl">80.56377</span> <span class="fl">81.29680</span> <span class="fl">80.52947</span> <span class="fl">80.10540</span> <span class="fl">80.41855</span> <span class="fl">80.58089</span></span>
<span id="cb8-32"><a href="#cb8-32" aria-hidden="true" tabindex="-1"></a>[<span class="dv">27</span>] <span class="fl">80.27750</span> <span class="fl">80.49915</span> <span class="fl">79.77512</span> <span class="fl">79.54571</span> <span class="fl">79.07448</span> <span class="fl">79.59706</span> <span class="fl">79.30831</span> <span class="fl">79.04140</span> <span class="fl">79.46599</span> <span class="fl">79.23228</span> <span class="fl">78.96974</span> <span class="fl">78.89496</span> <span class="fl">78.87805</span></span>
<span id="cb8-33"><a href="#cb8-33" aria-hidden="true" tabindex="-1"></a>[<span class="dv">40</span>] <span class="fl">79.21300</span> <span class="fl">79.62620</span> <span class="fl">79.49839</span> <span class="fl">79.30760</span> <span class="fl">79.69878</span></span>
<span id="cb8-34"><a href="#cb8-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-35"><a href="#cb8-35" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>GG</span>
<span id="cb8-36"><a href="#cb8-36" aria-hidden="true" tabindex="-1"></a>     [,<span class="dv">1</span>] [,<span class="dv">2</span>] [,<span class="dv">3</span>] [,<span class="dv">4</span>] [,<span class="dv">5</span>] [,<span class="dv">6</span>] [,<span class="dv">7</span>] [,<span class="dv">8</span>] [,<span class="dv">9</span>] [,<span class="dv">10</span>] [,<span class="dv">11</span>] [,<span class="dv">12</span>] [,<span class="dv">13</span>] [,<span class="dv">14</span>] [,<span class="dv">15</span>] [,<span class="dv">16</span>] [,<span class="dv">17</span>] [,<span class="dv">18</span>] [,<span class="dv">19</span>] [,<span class="dv">20</span>]</span>
<span id="cb8-37"><a href="#cb8-37" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,]   <span class="dv">11</span>   <span class="dv">12</span>   <span class="dv">13</span>   <span class="dv">14</span>   <span class="dv">15</span>   <span class="dv">16</span>   <span class="dv">17</span>   <span class="dv">18</span>   <span class="dv">19</span>    <span class="dv">20</span>    <span class="dv">21</span>    <span class="dv">22</span>    <span class="dv">23</span>    <span class="dv">24</span>    <span class="dv">25</span>    <span class="dv">26</span>    <span class="dv">27</span>    <span class="dv">28</span>    <span class="dv">29</span>    <span class="dv">30</span></span>
<span id="cb8-38"><a href="#cb8-38" aria-hidden="true" tabindex="-1"></a>     [,<span class="dv">21</span>] [,<span class="dv">22</span>] [,<span class="dv">23</span>] [,<span class="dv">24</span>] [,<span class="dv">25</span>] [,<span class="dv">26</span>] [,<span class="dv">27</span>] [,<span class="dv">28</span>] [,<span class="dv">29</span>] [,<span class="dv">30</span>] [,<span class="dv">31</span>] [,<span class="dv">32</span>] [,<span class="dv">33</span>] [,<span class="dv">34</span>] [,<span class="dv">35</span>] [,<span class="dv">36</span>] [,<span class="dv">37</span>] [,<span class="dv">38</span>] [,<span class="dv">39</span>]</span>
<span id="cb8-39"><a href="#cb8-39" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,]    <span class="dv">31</span>    <span class="dv">32</span>    <span class="dv">33</span>    <span class="dv">34</span>    <span class="dv">35</span>    <span class="dv">36</span>    <span class="dv">37</span>    <span class="dv">38</span>    <span class="dv">39</span>    <span class="dv">40</span>    <span class="dv">41</span>    <span class="dv">42</span>    <span class="dv">43</span>    <span class="dv">44</span>    <span class="dv">45</span>    <span class="dv">46</span>    <span class="dv">47</span>    <span class="dv">48</span>    <span class="dv">49</span></span>
<span id="cb8-40"><a href="#cb8-40" aria-hidden="true" tabindex="-1"></a>     [,<span class="dv">40</span>] [,<span class="dv">41</span>] [,<span class="dv">42</span>] [,<span class="dv">43</span>] [,<span class="dv">44</span>]</span>
<span id="cb8-41"><a href="#cb8-41" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,]    <span class="dv">50</span>    <span class="dv">51</span>    <span class="dv">52</span>    <span class="dv">53</span>    <span class="dv">54</span></span>
<span id="cb8-42"><a href="#cb8-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-43"><a href="#cb8-43" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(,<span class="st">"class"</span>)</span>
<span id="cb8-44"><a href="#cb8-44" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"paleoTSfit"</span></span></code></pre></div>
<p><br> The model with the highest likelihood has a punctuation event
at sample (population) 18. The rate of evolution before the punctuation
(<em>v_step.1</em>) is about twice the magnitude of the rate of
evolution after the punctuation (<em>v_step.2</em>).</p>
<p>Note: <code>all.logl</code> lists the log-likelihoods for all
investigated shift points. All investigated shift points are given by
<code>GG</code>.</p>
<p>The log-likelihood surface can be investigated using the
<code>confidence.PE.Bokma</code> function. Notice that we need to
condition the estimated log-likelihoods for the pairwise parameter
combinations of <em>vstep_1</em> and <em>vstep_2</em> on the most likely
parameter estimates of the punctuation (<em>shift1</em>), ancestral
trait value (<em>anc</em>) and the new optimum after the punctuation
(<em>new.opt</em>). <br></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">loglik.surface.Bokma</span> (ln.diameter, <span class="at">vstep_1.vec=</span><span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="fl">0.01</span>), <span class="at">vstep_2.vec=</span><span class="fu">seq</span>(<span class="dv">0</span>,<span class="fl">0.4</span>, <span class="fl">0.01</span>),<span class="at">shift=</span><span class="dv">18</span>, <span class="at">anc=</span><span class="fl">3.7157502</span>  , <span class="at">new.opt=</span><span class="fl">4.1349082</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>        lower upper</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>vstep_1  <span class="fl">0.14</span>  <span class="fl">0.91</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>vstep_2  <span class="fl">0.10</span>  <span class="fl">0.30</span></span></code></pre></div>
<p><br><img src="boka_logl_surface.png" width="60%" style="display: block; margin: auto;"><br> We can compare the fit of the Bokma model to the alternative
interpretation of a punctuated mode of evolution where the trait evolves
according to a model of stasis in between the punctuation events. This
model is implemented in the <code>paleoTS</code> package and can be
fitted using the <code>fitGpunc</code> function: <br></p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> Stasis.PE<span class="ot">&lt;-</span>paleoTS<span class="sc">::</span><span class="fu">fitGpunc</span>(ln.diameter, <span class="at">minb =</span> <span class="dv">10</span>, <span class="at">silent =</span> <span class="cn">TRUE</span>, <span class="at">oshare =</span> <span class="cn">FALSE</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> Bokma.PE<span class="sc">$</span>AICc;Stasis.PE<span class="sc">$</span>AICc</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">155.7136</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">114.9293</span></span></code></pre></div>
<p><br> The model where the trait evolves as a unbiased random walk in
between the punctuation event (i.e. the Bokma model) has the best
relative fit in this case. <br></p>
</div>
<div class="section level3">
<h3 id="ornstein-uhlenbeck-model-with-moving-optimum-">4.4 Ornstein-Uhlenbeck model with moving optimum.<a class="anchor" aria-label="anchor" href="#ornstein-uhlenbeck-model-with-moving-optimum-"></a>
</h3>
<p>An Ornstein-Uhlebeck model describe the evolution of a trait towards
an optimum. The <code>paleoTS</code> package includes an
Ornstein-Uhlenbeck (OU) model of evolution with a single, fixed optimum
(Hunt <em>et al.</em> 2008), i.e. it portrays evolutionary adaptation of
a trait towards a fixed peak on the adaptive landscape. However, peaks
in the adaptive landscape might not be fixed and the <code>evoTS</code>
package contains functions to fit OU models where the optimum (peak) is
constantly changing according to an unbiased random walk. Such a model
was proposed by Hansen <em>et al.</em> (2008) for analyses of
phylogenetic comparative data. Adjusted to describe evolution of a
single lineage, the expected trait mean and its variance and covariance
are given by the following expressions: <br><br><span class="math display">\[E[z_{i}] = e^{(-\alpha t_{i})}z_{0} +
[1-e^{-\alpha t_{i}}]\theta\]</span> <br><span class="math display">\[Var[z_{i}] =\left[ \frac{
\sigma^{2}_{step}+\sigma^{2}_{\theta}}{2\alpha} \right] \left[
1-e^{(-2\alpha t_{i})}\right] + \sigma^{2}_{\theta}t_{i}
\left[  1-2(1-e^{-\alpha t_{i}}) \right]/\alpha t_{i} + \epsilon _{i}
\]</span> <br><span class="math display">\[Cov[z_{i},z_{j}] =\left[
\frac{ \sigma^{2}_{step}+\sigma^{2}_{\theta}}{2\alpha} \right] \left[
1-e^{(-2\alpha t_{a})}\right]e^{-\alpha t_{ij}} +
\sigma^{2}_{\theta}t_{a} \left[  1-\left[1+e^{-\alpha t_{ij}} \right]
\left( 1-e^{-\alpha t_{a}} \right) \right] /  \alpha t_{i}\]</span>
<br> where <span class="math inline">\(z_{i}\)</span> is the expected
trait value for the <em>i</em>th sample, <span class="math inline">\(z_{0}\)</span> is the ancestral trait mean, <span class="math inline">\(t_{i}\)</span> is the time interval from the
ancestral population mean (the start of the time-series, which has a
time of 0) to the <em>i</em>th sample, <span class="math inline">\(\theta\)</span> is the optimum, $$ measures the
rate of adaptation to the optimum, <span class="math inline">\(\sigma^{2}_{step}\)</span> is the variance of the
stochastic perturbations of <em>z</em>, and <span class="math inline">\(\sigma^{2}_{\theta}\)</span> is the variance of
the stochastic perturbations of the optimum, <span class="math inline">\(t_{a}\)</span> is the time interval from the
ancestral population to the oldest of the two populations <span class="math inline">\(z_{i}\)</span> and <span class="math inline">\(z_{j}\)</span>, and <span class="math inline">\(t_{ij}\)</span> is the time separating two samples
<span class="math inline">\(z_{i}\)</span> and <span class="math inline">\(z_{j}\)</span>. The estimation (sampling) error
<span class="math inline">\(\epsilon_{i}\)</span> of the population
means contribute to the expected variance between two population
means.</p>
<p>The model can be fitted using the <code>opt.joint.OUBM</code>
function. <br></p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">opt.joint.OUBM</span>(ln.diameter)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Running no iterations."</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>logL</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">78.5667</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>AICc</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">148.4437</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>parameters</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>anc<span class="sc">/</span>theta<span class="fl">.0</span> vstep.trait       alpha   vstep.opt </span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a> <span class="fl">3.71050957</span>  <span class="fl">0.25577371</span>  <span class="fl">4.45009756</span>  <span class="fl">0.00000001</span> </span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>modelName</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"OU model with moving optimum (ancestral state at optimum)"</span></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>method</span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Joint"</span></span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>K</span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">4</span></span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>n</span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">63</span></span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-25"><a href="#cb11-25" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>iter</span>
<span id="cb11-26"><a href="#cb11-26" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="cn">NA</span></span>
<span id="cb11-27"><a href="#cb11-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-28"><a href="#cb11-28" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(,<span class="st">"class"</span>)</span>
<span id="cb11-29"><a href="#cb11-29" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"paleoTSfit"</span></span></code></pre></div>
<p><br> The <em>vstep.opt</em> parameter describes the rate of change
in the optimum. This is extremely small (virtually zero) in the example
above, which means the optimum is essentially fixed. The alpha in the OU
model represents the strength of the pull towards the optimum (Hansen
1997). A parameter that is easier to interpret compared to the alpha is
the half-life, <span class="math inline">\(ln(2)/ \alpha\)</span>, which
is the time it takes for the trait to move half-way from the ancestral
state to the optimum. The half life is therefore a quantification of the
speed of adaptation towards the optimal state. As for the decelerated
and accelerated models of evolution, the interpretation of the half life
depends on the time-interval covered by the time-series. Since the
time-interval of the time-series we analyze is scaled to unit length
(i.e. the time from the start to the end of the time-series is 1), this
means the half-life can be interpreted as the percent of the total
length of the time-series. The half-life in our example is <span class="math inline">\(ln(2)/ \alpha\)</span> = 0.16. According to this
point estimate, it takes the trait 16% of the total length of the
time-series to evolve half-way towards the optimum, which is about
(13,728 years * 0.16 =) 2197 years.</p>
<p>Note that the name of the first reported parameter is
<em>anc</em>/<em>theta.0</em>. This parameter represents the ancestral
trait value, but also the value of the “ancestral” optimum. The default
option in the <code>opt.joint.OUBM</code> function is to assume that the
trait was perfectly adapted at the start of the time-series (the
argument <code>anc.opt = TRUE</code>), but this can be changed by
setting <code>anc.opt = FALSE</code> <br></p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">opt.joint.OUBM</span>(ln.diameter, <span class="at">opt.anc  =</span> <span class="cn">FALSE</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Running no iterations."</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>logL</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">80.71298</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>AICc</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">150.3733</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>parameters</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>        anc vstep.trait     theta<span class="fl">.0</span>       alpha   vstep.opt </span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a> <span class="fl">3.70316688</span>  <span class="fl">0.27295686</span>  <span class="fl">3.89044533</span> <span class="fl">11.89309009</span>  <span class="fl">0.00000001</span> </span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>modelName</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"OU model with moving optimum"</span></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>method</span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Joint"</span></span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>K</span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">5</span></span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>n</span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">63</span></span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>iter</span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="cn">NA</span></span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(,<span class="st">"class"</span>)</span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"paleoTSfit"</span></span></code></pre></div>
<p><br> Setting <code>opt.anc  = FALSE</code> estimates a separate
“ancestral” value for the optimum (<em>theta.0</em>). The rate of change
in the optimum (<code>vstep.opt</code>) is still negligible, which means
this model is virtually identical to a model where the optimum is fixed.
This can be shown by fitting an OU model where the optimum is fixed,
which is the model included in the <code>paleoTS</code> package.
<br></p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> paleoTS<span class="sc">::</span><span class="fu">opt.joint.OU</span>(ln.diameter)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>logL</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">80.71298</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>AICc</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">152.7363</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>parameters</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>       anc      vstep      theta      alpha </span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a> <span class="fl">3.7031735</span>  <span class="fl">0.2729604</span>  <span class="fl">3.8904516</span> <span class="fl">11.8941703</span> </span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>modelName</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"OU"</span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>method</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Joint"</span></span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>se</span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a><span class="cn">NULL</span></span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>K</span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">4</span></span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>n</span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">63</span></span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(,<span class="st">"class"</span>)</span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"paleoTSfit"</span></span></code></pre></div>
<p><br> The fixed optimum model gives the same log-likelihood value as
the model where the optimum was allowed to change (but actually didn’t).
The fixed optimum model has a better AICc score as this model contains
one less parameter (the parameter describing the rate of change in the
optimum).</p>
<p>It is good practice to repeat any numerical optimization procedure
from different starting points. This is especially important when the
model has several parameters, as parameter-rich models may contain more
than one peak in the log-likelihood surface. The OUBM model is a type of
model that may have several local peaks in the likelihood space.</p>
<p>The user can choose the number of iterations of the numerical
optimization of the OUBM model using the argument
<code>iterations.</code> The function will return the parameter values
from the run with the highest log-likelihood. The starting values in
each iteration are drawn from a normal distribution with mean zero and a
standard deviation set by the user (default is 1). The initial values
for the <em>vstep</em> and <em>alpha</em> parameters are constrained to
be equal or larger than 0.</p>
<p>Here, we run the <code>opt.joint.OUBM</code> function (assuming the
trait value is perfectly adapted at the start of the sequence) from 100
different starting points (i.e. different initial parameter values):
<br></p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">opt.joint.OUBM</span>(ln.diameter, <span class="at">opt.anc =</span> <span class="cn">TRUE</span>, <span class="at">iterations =</span> <span class="dv">100</span>)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>logL</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">78.5667</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>AICc</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">148.4437</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>parameters</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>anc<span class="sc">/</span>theta<span class="fl">.0</span> vstep.trait       alpha   vstep.opt </span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a> <span class="fl">3.71050959</span>  <span class="fl">0.25577389</span>  <span class="fl">4.45010533</span>  <span class="fl">0.00000001</span> </span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>modelName</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"OU w. moving optimum (anc at opt)"</span></span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>method</span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Joint"</span></span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>K</span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">4</span></span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>n</span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">63</span></span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>iter</span>
<span id="cb14-25"><a href="#cb14-25" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">100</span></span>
<span id="cb14-26"><a href="#cb14-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-27"><a href="#cb14-27" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(,<span class="st">"class"</span>)</span>
<span id="cb14-28"><a href="#cb14-28" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"paleoTSfit"</span></span></code></pre></div>
<p><br> From the output, we see that the likelihood score of the best
model among the 100 model runs is identical to the score when we ran the
model without any iterations. However, the maximum likelihood parameter
estimates are slightly different (e.g. a difference in the sixth decimal
for the <em>vstep</em> parameter), but not to an extent that changes our
interpretation of the trait dynamics. This suggests that the detected
peak in the log-likelihood landscape is not completely smooth, and that
there exists minuscule spikes very close to each other at the peak.</p>
<p>The <code>evoTS</code> package contains functions to estimate
likelihood surfaces for the different versions of the OU models
(<code>confidence.OU</code> and <code>confidence.OUBM</code>). In these
functions, the likelihood surface is not estimated as a function of the
step variance and alpha parameter directly, but rather as a function of
two related parameters that are easier to give a biological
interpretation. The stationary variance, <span class="math inline">\(vstep/(2* \alpha)\)</span>, represents the
equilibrium variance of the OU process (Hansen <em>et al.</em> 2008) and
describes the variance expected in the trait after the trait has reached
the optimum. The half-life, <span class="math inline">\(log(2)/(\alpha)\)</span>, is the amount of time it
takes for the trait to move half-way from the ancestral state to the
optimum. The half-life is informative regarding the speed of adaptation
toward the optimal state.</p>
<p>The OU model with a fixed optimum had the best relative model fit
according to AICc among the three versions of the OU model we
investigated. The half-life from this OU-model is <span class="math inline">\(log(2)/11.8941\)</span> = 0.0583. Since we scaled
the time-interval of the time-series to unit length (i.e. the time from
the start to the end of the time-series is 1), this means it takes about
6% of the length of our data to get half-way to the optimal value. The
total length of our time-series in years is 13,728, which means it takes
(13728 * 0.0583 =) 800 years to get half-way to the optimum. But this is
only a point-estimate. We can explore the support interval around this
point estimate of the half-life (and the stationary variance) using the
<code>confidence.OU</code> function. <br></p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">loglik.surface.OU</span>(ln.diameter, <span class="at">stat.var.vec=</span><span class="fu">seq</span>(<span class="dv">0</span>,<span class="fl">0.1</span>,<span class="fl">0.001</span>), <span class="at">h.vec=</span><span class="fu">seq</span>(<span class="dv">0</span>,<span class="fl">0.4</span>,<span class="fl">0.001</span>))</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>      lower upper</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>stationary variance <span class="fl">0.007</span> <span class="fl">0.053</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>half<span class="sc">-</span>life           <span class="fl">0.029</span> <span class="fl">0.305</span></span></code></pre></div>
<p><br><img src="OU_logl_surface.png" width="60%" style="display: block; margin: auto;"><br></p>
<p>Half-life values up to 30% of the total length of the time-series are
within two log-likelihood units from the best estimate. This indicates
that substantially slower evolution than the point estimate of 6% cannot
be ruled out as a possibility.</p>
</div>
<div class="section level3">
<h3 id="fitting-all-univariate-models-in-evots-and-paleots">4.5 Fitting all univariate models in <code>evoTS</code> and
<code>paleoTS</code><a class="anchor" aria-label="anchor" href="#fitting-all-univariate-models-in-evots-and-paleots"></a>
</h3>
<p>A quick way to evaluate the relative fit of all univariate models in
the <code>evoTS</code> and <code>paleoTS</code>packages (excluding
models with more than one punctuation/mode shift) is to use the
<code>fit.all.univariate.models</code> function. <br></p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">fit.all.univariate.models</span>(ln.diameter)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Omitting some of the output to save length. </span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>Comparing <span class="dv">11</span> models [n <span class="ot">=</span> <span class="dv">63</span>, method <span class="ot">=</span> Joint]</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>                                                                logL K       AICc       dAICc Akaike.wt</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>GRW                                                         <span class="fl">77.64073</span> <span class="dv">3</span> <span class="sc">-</span><span class="fl">148.87469</span>    <span class="fl">6.838905</span>     <span class="fl">0.022</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>URW                                                         <span class="fl">77.57018</span> <span class="dv">2</span> <span class="sc">-</span><span class="fl">150.94035</span>    <span class="fl">4.773241</span>     <span class="fl">0.063</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>Stasis                                                      <span class="fl">39.84019</span> <span class="dv">2</span>  <span class="sc">-</span><span class="fl">75.48039</span>   <span class="fl">80.233205</span>     <span class="fl">0.000</span></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>StrictStasis                                              <span class="sc">-</span><span class="fl">707.46411</span> <span class="dv">1</span> <span class="fl">1416.99379</span> <span class="fl">1572.707381</span>     <span class="fl">0.000</span></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>Decel                                                       <span class="fl">76.77642</span> <span class="dv">3</span> <span class="sc">-</span><span class="fl">147.14607</span>    <span class="fl">8.567527</span>     <span class="fl">0.009</span></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>Accel                                                       <span class="fl">73.80287</span> <span class="dv">3</span> <span class="sc">-</span><span class="fl">141.19896</span>   <span class="fl">14.514630</span>     <span class="fl">0.000</span></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>OU                                                          <span class="fl">80.71298</span> <span class="dv">4</span> <span class="sc">-</span><span class="fl">152.73631</span>    <span class="fl">2.977287</span>     <span class="fl">0.155</span></span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>OU model with moving <span class="fu">optimum</span> (ancestral state at optimum)   <span class="fl">78.56670</span> <span class="dv">4</span> <span class="sc">-</span><span class="fl">148.44374</span>    <span class="fl">7.269854</span>     <span class="fl">0.018</span></span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>OU model with moving optimum                                <span class="fl">80.71298</span> <span class="dv">5</span> <span class="sc">-</span><span class="fl">150.37333</span>    <span class="fl">5.340265</span>     <span class="fl">0.047</span></span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>Punc<span class="dv">-1</span>                                                      <span class="fl">50.03827</span> <span class="dv">4</span>  <span class="sc">-</span><span class="fl">91.38688</span>   <span class="fl">64.326713</span>     <span class="fl">0.000</span></span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>Bokma model                                                 <span class="fl">83.38311</span> <span class="dv">5</span> <span class="sc">-</span><span class="fl">155.71359</span>    <span class="fl">0.000000</span>     <span class="fl">0.685</span></span></code></pre></div>
<p><br> The Bokma model out competes the OU model according to
AICc.</p>
<p>Where the jump happens can be investigated by plotting the data and
putting a line at the estimated switch point.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">plotpaleoTS</span>(ln.diameter)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">abline</span>(<span class="at">v=</span>ln.diameter<span class="sc">$</span>tt[<span class="dv">18</span>], <span class="at">col=</span><span class="st">"red"</span>, <span class="at">lty=</span><span class="dv">2</span>, <span class="at">lwd=</span><span class="dv">2</span>)</span></code></pre></div>
<p><br><br><img src="diameter_jump.png" width="70%" style="display: block; margin: auto;"><br><br> The switch point is estimated to happen where there is no
obvious and large jump in the evolution of the trait along the y axis.
The reason why the Bokma model fits better than the other candidate
models we have investigated might therefore be unrelated to the jump
itself and may instead be due to the fact that the rate of evolution is
allowed to be different before and after the punctuation. The rate
parameter of the unbiased random walk before the punctuation
(<em>vstep_1</em>) in the Bokma model is indeed about twice as large as
the rate parameter in the unbiased random walk after the punctuation
(<em>vstep_2</em>). <br></p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> Bokma.PE<span class="sc">$</span>parameters</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>       anc    vstep_1    new.opt    vstep_2     shift1 </span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a> <span class="fl">3.7157496</span>  <span class="fl">0.3419321</span>  <span class="fl">4.1349097</span>  <span class="fl">0.1668655</span> <span class="fl">18.0000000</span>  </span></code></pre></div>
<p><br> A model that fits two separate unbiased random walks to two
non-overlapping segments of the time-series, but where the mode shift
does not involve a jump in phenotype space, might therefore out compete
the Bokma model.</p>
</div>
<div class="section level3">
<h3 id="fitting-combinations-of-univariate-models-to-a-time-series-mode-shift">4.6 Fitting combinations of univariate models to a time series (mode
shift)<a class="anchor" aria-label="anchor" href="#fitting-combinations-of-univariate-models-to-a-time-series-mode-shift"></a>
</h3>
<p>There is no a priori reason why a lineage should follow only one mode
of evolution. In addition to the Bokma model where a mode shift is
associated with a jump in phenotype space (punctuation), the
<code>evoTS</code> package allows for investigating all pairwise model
combinations of the models stasis, unbiased random walk (URW), trend
(GRW) and Ornstein-Uhlenbeck (OU) process with a fixed optimum using the
function <code>fit.mode.shift</code>.</p>
<p>We are interested in investigating whether a model consisting of two
different unbiased random walks fits the data better compared to a
punctuated evolution model (The Bokma model). We can constrain the
function to only investigate the same shift point as the one found to be
the best for the Bokma model using the argument
<code>shift.point</code>. However, we are interested in investigating
all possible switch points. <br></p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">fit.mode.shift</span>(ln.diameter, <span class="at">model1 =</span> <span class="st">"URW"</span>, <span class="at">model2 =</span> <span class="st">"URW"</span>, <span class="at">minb =</span> <span class="dv">10</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Searching all possible switchpoints in the evolutionary sequence"</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>Total <span class="co"># hypotheses:  44 </span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>  <span class="dv">2</span>  <span class="dv">3</span>  <span class="dv">4</span>  <span class="dv">5</span>  <span class="dv">6</span>  <span class="dv">7</span>  <span class="dv">8</span>  <span class="dv">9</span>  <span class="dv">10</span>  <span class="dv">11</span>  <span class="dv">12</span>  <span class="dv">13</span>  <span class="dv">14</span>  <span class="dv">15</span>  <span class="dv">16</span>  <span class="dv">17</span>  <span class="dv">18</span>  <span class="dv">19</span>  <span class="dv">20</span>  <span class="dv">21</span>  <span class="dv">22</span>  <span class="dv">23</span>  <span class="dv">24</span>  <span class="dv">25</span>  <span class="dv">26</span>  <span class="dv">27</span>  <span class="dv">28</span>  <span class="dv">29</span>  <span class="dv">30</span>  <span class="dv">31</span>  <span class="dv">32</span>  <span class="dv">33</span>  <span class="dv">34</span>  <span class="dv">35</span>  <span class="dv">36</span>  <span class="dv">37</span>  <span class="dv">38</span>  <span class="dv">39</span>  <span class="dv">40</span>  <span class="dv">41</span>  <span class="dv">42</span>  <span class="dv">43</span>  <span class="dv">44</span>  </span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>logL</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">79.27473</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>AICc</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">149.8598</span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>parameters</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>       anc      vstep      vstep     shift1 </span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a> <span class="fl">3.7304865</span>  <span class="fl">0.2432606</span>  <span class="fl">0.2494830</span> <span class="fl">52.0000000</span> </span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>modelName</span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"URW-URW"</span></span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>method</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Joint"</span></span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>se</span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a><span class="cn">NULL</span></span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>K</span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">4</span></span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>n</span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">63</span></span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-30"><a href="#cb19-30" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>all.logl</span>
<span id="cb19-31"><a href="#cb19-31" aria-hidden="true" tabindex="-1"></a> [<span class="dv">1</span>] <span class="fl">70.86731</span> <span class="fl">68.59747</span> <span class="fl">62.65156</span> <span class="fl">65.41455</span> <span class="fl">53.23718</span> <span class="fl">60.25449</span> <span class="fl">56.19608</span> <span class="fl">45.50594</span> <span class="fl">42.55632</span> <span class="fl">44.70092</span> <span class="fl">46.05836</span> <span class="fl">46.58797</span> <span class="fl">48.14004</span></span>
<span id="cb19-32"><a href="#cb19-32" aria-hidden="true" tabindex="-1"></a>[<span class="dv">14</span>] <span class="fl">44.19647</span> <span class="fl">67.63474</span> <span class="fl">57.82010</span> <span class="fl">64.69928</span> <span class="fl">69.73926</span> <span class="fl">58.45089</span> <span class="fl">55.26295</span> <span class="fl">53.40897</span> <span class="fl">64.95625</span> <span class="fl">66.97684</span> <span class="fl">66.65108</span> <span class="fl">72.98220</span> <span class="fl">65.42928</span></span>
<span id="cb19-33"><a href="#cb19-33" aria-hidden="true" tabindex="-1"></a>[<span class="dv">27</span>] <span class="fl">71.12637</span> <span class="fl">76.39216</span> <span class="fl">74.96141</span> <span class="fl">74.07654</span> <span class="fl">74.06923</span> <span class="fl">77.98640</span> <span class="fl">75.27489</span> <span class="fl">73.96364</span> <span class="fl">78.18181</span> <span class="fl">74.92223</span> <span class="fl">75.10229</span> <span class="fl">77.07127</span> <span class="fl">77.28518</span></span>
<span id="cb19-34"><a href="#cb19-34" aria-hidden="true" tabindex="-1"></a>[<span class="dv">40</span>] <span class="fl">79.12256</span> <span class="fl">75.99980</span> <span class="fl">79.27473</span> <span class="fl">76.45725</span> <span class="fl">70.37393</span></span>
<span id="cb19-35"><a href="#cb19-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-36"><a href="#cb19-36" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>GG</span>
<span id="cb19-37"><a href="#cb19-37" aria-hidden="true" tabindex="-1"></a>shift1 </span>
<span id="cb19-38"><a href="#cb19-38" aria-hidden="true" tabindex="-1"></a>    <span class="dv">52</span> </span>
<span id="cb19-39"><a href="#cb19-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-40"><a href="#cb19-40" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(,<span class="st">"class"</span>)</span>
<span id="cb19-41"><a href="#cb19-41" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"paleoTSfit"</span></span></code></pre></div>
<p><br> The fit of this model is similar (but worse) according to AICc
compared to the Bokma model.</p>
<p>It is also possible to use the function <code>fit.mode.shift</code>
to fit all pairwise combinations of the four models by setting the
<code>fit.all</code> argument as <code>TRUE.</code> If a shift point is
not defined (using the <code>shift.point</code> argument), all possible
shift points are investigated for all models. <br></p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">fit.mode.shift</span>(ln.diameter, <span class="at">fit.all =</span> <span class="cn">TRUE</span>, <span class="at">minb =</span> <span class="dv">10</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Searching all possible switchpoints in the evolutionary sequence"</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>  <span class="dv">2</span>  <span class="dv">3</span>  <span class="dv">4</span>  <span class="dv">5</span>  <span class="dv">6</span>  <span class="dv">7</span>  <span class="dv">8</span>  <span class="dv">9</span>  <span class="dv">10</span>  <span class="dv">11</span>  <span class="dv">12</span>  <span class="dv">13</span>  <span class="dv">14</span>  <span class="dv">15</span>  <span class="dv">16</span>  <span class="dv">17</span>  <span class="dv">18</span>  <span class="dv">19</span>  <span class="dv">20</span>  <span class="dv">21</span>  <span class="dv">22</span>  <span class="dv">23</span>  <span class="dv">24</span>  <span class="dv">25</span>  <span class="dv">26</span>  <span class="dv">27</span>  <span class="dv">28</span>  <span class="dv">29</span>  <span class="dv">30</span>  <span class="dv">31</span>  <span class="dv">32</span>  <span class="dv">33</span>  <span class="dv">34</span>  <span class="dv">35</span>  <span class="dv">36</span>  <span class="dv">37</span>  <span class="dv">38</span>  <span class="dv">39</span>  <span class="dv">40</span>  <span class="dv">41</span>  <span class="dv">42</span>  <span class="dv">43</span>  <span class="dv">44</span>  </span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>Comparing <span class="dv">16</span> models [n <span class="ot">=</span> <span class="dv">63</span>, method <span class="ot">=</span> Joint]</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>                  logL K       AICc      dAICc Akaike.wt</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>Stasis<span class="sc">-</span>Stasis <span class="fl">53.24711</span> <span class="dv">4</span>  <span class="sc">-</span><span class="fl">97.80456</span> <span class="fl">64.3140156</span>     <span class="fl">0.000</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>Stasis<span class="sc">-</span>URW    <span class="fl">72.84306</span> <span class="dv">4</span> <span class="sc">-</span><span class="fl">136.99646</span> <span class="fl">25.1221145</span>     <span class="fl">0.000</span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>Stasis<span class="sc">-</span>GRW    <span class="fl">72.84414</span> <span class="dv">5</span> <span class="sc">-</span><span class="fl">134.63564</span> <span class="fl">27.4829349</span>     <span class="fl">0.000</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>Stasis<span class="sc">-</span>OU     <span class="fl">74.97027</span> <span class="dv">6</span> <span class="sc">-</span><span class="fl">136.44053</span> <span class="fl">25.6780478</span>     <span class="fl">0.000</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>URW<span class="sc">-</span>URW       <span class="fl">79.27473</span> <span class="dv">4</span> <span class="sc">-</span><span class="fl">149.85981</span> <span class="fl">12.2587671</span>     <span class="fl">0.001</span></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>URW<span class="sc">-</span>GRW       <span class="fl">79.54430</span> <span class="dv">5</span> <span class="sc">-</span><span class="fl">148.03597</span> <span class="fl">14.0826065</span>     <span class="fl">0.000</span></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>URW<span class="sc">-</span>OU        <span class="fl">85.34081</span> <span class="dv">6</span> <span class="sc">-</span><span class="fl">157.18163</span>  <span class="fl">4.9369506</span>     <span class="fl">0.026</span></span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>GRW<span class="sc">-</span>GRW       <span class="fl">84.03615</span> <span class="dv">6</span> <span class="sc">-</span><span class="fl">154.57229</span>  <span class="fl">7.5462836</span>     <span class="fl">0.007</span></span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>GRW<span class="sc">-</span>OU        <span class="fl">88.87002</span> <span class="dv">7</span> <span class="sc">-</span><span class="fl">161.70368</span>  <span class="fl">0.4148986</span>     <span class="fl">0.254</span></span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>OU<span class="sc">-</span>OU         <span class="fl">90.26555</span> <span class="dv">8</span> <span class="sc">-</span><span class="fl">161.86443</span>  <span class="fl">0.2541469</span>     <span class="fl">0.275</span></span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a>OU<span class="sc">-</span>GRW        <span class="fl">84.14765</span> <span class="dv">7</span> <span class="sc">-</span><span class="fl">152.25894</span>  <span class="fl">9.8596410</span>     <span class="fl">0.002</span></span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>OU<span class="sc">-</span>URW        <span class="fl">83.90670</span> <span class="dv">6</span> <span class="sc">-</span><span class="fl">154.31340</span>  <span class="fl">7.8051743</span>     <span class="fl">0.006</span></span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>OU<span class="sc">-</span>Stasis     <span class="fl">87.80929</span> <span class="dv">6</span> <span class="sc">-</span><span class="fl">162.11858</span>  <span class="fl">0.0000000</span>     <span class="fl">0.312</span></span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a>GRW<span class="sc">-</span>URW       <span class="fl">83.79338</span> <span class="dv">5</span> <span class="sc">-</span><span class="fl">156.53412</span>  <span class="fl">5.5844543</span>     <span class="fl">0.019</span></span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a>GRW<span class="sc">-</span>Stasis    <span class="fl">86.49922</span> <span class="dv">6</span> <span class="sc">-</span><span class="fl">159.49844</span>  <span class="fl">2.6201411</span>     <span class="fl">0.084</span></span>
<span id="cb20-22"><a href="#cb20-22" aria-hidden="true" tabindex="-1"></a>URW<span class="sc">-</span>Stasis    <span class="fl">83.36672</span> <span class="dv">5</span> <span class="sc">-</span><span class="fl">155.68080</span>  <span class="fl">6.4377793</span>     <span class="fl">0.012</span></span>
<span id="cb20-23"><a href="#cb20-23" aria-hidden="true" tabindex="-1"></a>[[<span class="dv">1</span>]]</span>
<span id="cb20-24"><a href="#cb20-24" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>logL</span>
<span id="cb20-25"><a href="#cb20-25" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">87.80929</span></span>
<span id="cb20-26"><a href="#cb20-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-27"><a href="#cb20-27" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>AICc</span>
<span id="cb20-28"><a href="#cb20-28" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">162.1186</span></span>
<span id="cb20-29"><a href="#cb20-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-30"><a href="#cb20-30" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>parameters</span>
<span id="cb20-31"><a href="#cb20-31" aria-hidden="true" tabindex="-1"></a>         anc        vstep     theta_OU        alpha        omega       shift1 </span>
<span id="cb20-32"><a href="#cb20-32" aria-hidden="true" tabindex="-1"></a> <span class="fl">3.705695031</span>  <span class="fl">0.327387632</span>  <span class="fl">3.817864883</span>  <span class="fl">5.226471406</span>  <span class="fl">0.001870449</span> <span class="fl">38.000000000</span> </span>
<span id="cb20-33"><a href="#cb20-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-34"><a href="#cb20-34" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>modelName</span>
<span id="cb20-35"><a href="#cb20-35" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"OU-Stasis"</span></span>
<span id="cb20-36"><a href="#cb20-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-37"><a href="#cb20-37" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>method</span>
<span id="cb20-38"><a href="#cb20-38" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Joint"</span></span>
<span id="cb20-39"><a href="#cb20-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-40"><a href="#cb20-40" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>se</span>
<span id="cb20-41"><a href="#cb20-41" aria-hidden="true" tabindex="-1"></a><span class="cn">NULL</span></span>
<span id="cb20-42"><a href="#cb20-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-43"><a href="#cb20-43" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>K</span>
<span id="cb20-44"><a href="#cb20-44" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">6</span></span>
<span id="cb20-45"><a href="#cb20-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-46"><a href="#cb20-46" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>n</span>
<span id="cb20-47"><a href="#cb20-47" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">63</span></span>
<span id="cb20-48"><a href="#cb20-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-49"><a href="#cb20-49" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>GG</span>
<span id="cb20-50"><a href="#cb20-50" aria-hidden="true" tabindex="-1"></a>shift1 </span>
<span id="cb20-51"><a href="#cb20-51" aria-hidden="true" tabindex="-1"></a>    <span class="dv">38</span> </span>
<span id="cb20-52"><a href="#cb20-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-53"><a href="#cb20-53" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(,<span class="st">"class"</span>)</span>
<span id="cb20-54"><a href="#cb20-54" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"paleoTSfit"</span></span></code></pre></div>
<p><br> The function returns a list of the highest log-likelihood
found for each investigated model. A detailed output from the model with
the lowest AICc value among the 16 candidate models is also given. An
OU-Stasis model with a shift point at sample (population) 38 has the
best relative fit according to AICc. Note, however, that the
model-combination Trend-OU (GRW-OU) has an almost equal AICc score
relative to the best model. Also the combination of OU models (each with
their own fixed optimum) shows a good relative fit to the data. All
these three models have a better AICc score compared to the Bokma
model.</p>
</div>
<div class="section level3">
<h3 id="models-with-two-mode-shifts">4.7 Models with two mode shifts<a class="anchor" aria-label="anchor" href="#models-with-two-mode-shifts"></a>
</h3>
<p><code>evoTS</code> contains functions to fit some models with two
mode shifts. The <code>fit.2.Bokma</code> function fits three unbiased
random walks separated by two punctuation events. Other functions for
fitting two mode shifts focus on data sets where a “punctuation event”
(i.e. trend) is sampled and flanked by samples that may follow different
modes of evolution. These functions are
<code>opt.joint.Stasis.trend.Stasis</code>,
<code>opt.joint.Stasis.trend.URW</code>,
<code>opt.joint.URW.trend.Stasis</code>,
<code>opt.joint.URW.trend.URW</code>, and
<code>opt.joint.URW.URW.URW</code>.</p>
</div>
<div class="section level3">
<h3 id="simulating-data">4.8 Simulating data<a class="anchor" aria-label="anchor" href="#simulating-data"></a>
</h3>
<p>It is possible to simulate data for all implemented models in
<code>evoTS</code>. Standard R-package documentation can be seen by
entering <code><a href="../reference/sim.Bokma.html">?sim.Bokma</a></code>, <code><a href="../reference/sim.OUBM.html">?sim.OUBM</a></code> and
<code><a href="../reference/sim.accel.decel.html">?sim.accel.decel</a></code> <br><br><br></p>
</div>
</div>
<div class="section level2">
<h2 id="multivariate-models">5.0 Multivariate models<a class="anchor" aria-label="anchor" href="#multivariate-models"></a>
</h2>
<p>Traits are rarely changing independently of each other due to shared
genetics or development. Evolution is accordingly a multivariate
phenomenon. The <code>evoTS</code> package includes functions to fit the
following multivariate trait models:</p>
<ul>
<li><p>Multivariate unbiased random walks</p></li>
<li><p>Multivariate decelerated evolution</p></li>
<li><p>Multivariate accelerated evolution</p></li>
<li><p>Multivariate Ornstein-Uhlenbeck processes</p></li>
</ul>
<div class="section level3">
<h3 id="multivariate-unbiased-random-walks-with-and-without-changes-in-the-rate-of-evolution">5.1 Multivariate unbiased random walks with and without changes in
the rate of evolution<a class="anchor" aria-label="anchor" href="#multivariate-unbiased-random-walks-with-and-without-changes-in-the-rate-of-evolution"></a>
</h3>
<p>Evolution as a multivariate unbiased random walk is modeled using an
evolutionary rate matrix <strong>R</strong> (Felsenstein 1988; Revell
and Harmon 2008). The diagonal elements in <strong>R</strong> represent
the rate of evolution for the individual traits, while the off-diagonal
elements represent the extent to which different traits co-evolve.</p>
<p>The multivariate variance-covariance matrix for the unbiased random
walk model (<strong>V</strong>) is computed using the Kronecker product
of the <strong>R</strong> matrix and a “distance matrix”
<strong>C</strong>, describing the distance in time between the
different samples/populations in the time-series. Sampling error of the
trait mean (calculated as the sample variance divided by the sample
size) is – as for all models in <code>evoTS</code> – added to the
diagonal of the <strong>V</strong> matrix. To ensure symmetric positive
definiteness of the <strong>V</strong> matrix during log-likelihood
optimization, <strong>R</strong> is parameterized by its Cholesky
decomposition as the cross-product of upper triangular matrices. In
cases where different parts of the evolutionary time-series are
described by <span class="math inline">\(R_{m}\)</span> matrices,
<strong>V</strong> is computed as the sum of the different <span class="math inline">\(R_{m}\)</span> and <span class="math inline">\(C_{m}\)</span> matrices: <br><br><span class="math display">\[V = \sum_{i=1}^{m} R_{m} \otimes C_{m}\]</span>
<br> The current implementation of the multivariate unbiased random
walk model allows testing six variants of the model. All variants of the
model can be fitted using different specifications of <code>R</code> and
<code>r</code> in the <code>fit.multivariate.URW</code> function.</p>
<p>The <code>R</code> option defines which elements in the
<strong>R</strong> matrix that will be estimated. There are two options
for the structure of the <strong>R</strong> matrix. Setting
<code>R = "diagonal"</code> means only the diagonal elements of the
<strong>R</strong> matrix will be estimated while off-diagonal elements
are set to 0 (see panel a below) This parameterization of the
<strong>R</strong> matrix means the changes in the traits are assumed to
be uncorrelated. Setting <code>R = "symmetric"</code> means all (both
diagonal and off-diagonal) elements in the <strong>R</strong> matrix are
estimated (panel b below). This parameterization means the changes in
the traits are assumed to be correlated, i.e. that evolutionary changes
are non-independent. <br><br><img src="R_matrices.png" width="40%" style="display: block; margin: auto;"><br><br> The argument <code>r</code> in the
<code>fit.multivariate.URW</code> function defines whether the rate of
change is assumed constant (“fixed”), asymptotically decreasing
(“decel”), or asymptotically increasing (“accel”) with time. Defining
<code>r</code> as “fixed” means a regular multivariate unbiased random
walk will be fitted to the data. The “decel” and “accel” options fit
multivariate versions of the decelerated and accelerated versions of the
unbiased random walk, respectively. These latter two models deviate from
the multivariate unbiased random walk in that the distance matrix
<strong>C</strong> is transformed by an exponential decay or
acceleration parameter, <em>r</em>, that is jointly estimated during the
maximum likelihood search for the <strong>R</strong> matrix. A common
<em>r</em> parameter is assumed for all the traits.</p>
<p>To use the multivariate models, we first need multivariate data
(e.g. at least two traits or variables).</p>
<p>The data set on phenotypic evolution in <em>Stephanodiscus
yellowstonensis</em> (Theriot et al. 2006) contains data on the number
of ribs in addition to the size of the diameter we have analyzed so far.
Before combining the rib and diameter data into a multivariate data set,
we first do an approximate log-transformation of the rib data and
convert the time vector to unit length. <br></p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Doing an approximate log-transformation of the data</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> ln.ribs<span class="ot">&lt;-</span>paleoTS<span class="sc">::</span><span class="fu">ln.paleoTS</span>(ribs_S.yellowstonensis)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Convert the time vector to unit length</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> ln.ribs<span class="sc">$</span>tt<span class="ot">&lt;-</span>ln.ribs<span class="sc">$</span>tt<span class="sc">/</span>(<span class="fu">max</span>(ln.ribs<span class="sc">$</span>tt))</span></code></pre></div>
<p><br> To make a multivariate data set ready to be analyzed in
<code>evoTS</code>, we combine the two <code>paleoTS</code> objects into
a multivariate <code>evoTS</code> object using the
<code>make.multivar.evoTS</code> function. <br></p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> diam.ln_ribs.ln<span class="ot">&lt;-</span><span class="fu">make.multivar.evoTS</span>(ln.diameter, ln.ribs)</span></code></pre></div>
<p><br> We can use the <code>plotevoTS.multivariate</code> function to
have a look at the combined data set. <br></p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">plotevoTS.multivariate</span>(diam.ln_ribs.ln, <span class="at">y_min=</span><span class="fl">3.4</span>, <span class="at">y_max=</span><span class="fl">4.8</span>, <span class="at">x.label =</span> <span class="st">"Relative time"</span>, <span class="at">pch=</span><span class="fu">c</span>(<span class="dv">20</span>,<span class="dv">20</span>))</span></code></pre></div>
<p><br><img src="diam_and_ribs.png" width="60%" style="display: block; margin: auto;"><br><br> Eye-balling the data seems to suggest that the traits
change in a coordinated fashion.</p>
<p>We first fit a multivariate unbiased random walk model where the
off-diagonal elements in the <strong>R</strong> matrix are zero, and the
rate of evolution is assumes constant. This is equivalent as fitting to
separate univariate unbiased random walks models to each of the two
time-series. <br></p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">fit.multivariate.URW</span>(diam.ln_ribs.ln, <span class="at">R =</span> <span class="st">"diag"</span>, <span class="at">r =</span> <span class="st">"fixed"</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Model converged successfully."</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>modelName</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Multivariate Random walk (R matrix with zero off-diagonal elements)"</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>logL</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">135.8697</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>AICc</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">263.0498</span></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>ancestral.values</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">3.715649</span> <span class="fl">4.013975</span></span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>R</span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>          [,<span class="dv">1</span>]      [,<span class="dv">2</span>]</span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,] <span class="fl">0.2555386</span> <span class="fl">0.0000000</span></span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,] <span class="fl">0.0000000</span> <span class="fl">0.4695467</span></span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>SE.R</span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="cn">NA</span></span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>method</span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Joint"</span></span>
<span id="cb24-25"><a href="#cb24-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-26"><a href="#cb24-26" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>K</span>
<span id="cb24-27"><a href="#cb24-27" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">4</span></span>
<span id="cb24-28"><a href="#cb24-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-29"><a href="#cb24-29" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>n</span>
<span id="cb24-30"><a href="#cb24-30" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">63</span></span>
<span id="cb24-31"><a href="#cb24-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-32"><a href="#cb24-32" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>iter</span>
<span id="cb24-33"><a href="#cb24-33" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="cn">NA</span></span>
<span id="cb24-34"><a href="#cb24-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-35"><a href="#cb24-35" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(,<span class="st">"class"</span>)</span>
<span id="cb24-36"><a href="#cb24-36" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"paleoTSfit"</span></span></code></pre></div>
<p><br> The returned parameters include the ancestral trait values for
the two traits and the evolutionary rate matrix <strong>R</strong>. The
diagonal in the <strong>R</strong> matrix contains the step size (rate
of evolution) parameters. The first trait (diameter) has about twice the
rate of evolution as the second parameter (ribs). The off-diagonal
elements are zero as this model is not estimating the covariance of the
evolutionary changes in the two traits.</p>
<p>Next, we fit a model that allows the off-diagonal elements in the R
matrix to be different from zero. We do this by setting
<code>R = symmetric</code>. We are still keeping the rate of change
fixed through time. <br></p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">fit.multivariate.URW</span>(diam.ln_ribs.ln, <span class="at">R =</span> <span class="st">"symmetric"</span>, <span class="at">r =</span> <span class="st">"fixed"</span>)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Model converged successfully."</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>modelName</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Multivariate Random walk (R matrix with non-zero off-diagonal elements)"</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>logL</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">181.8526</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>AICc</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">352.6525</span></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>ancestral.values</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">3.719176</span> <span class="fl">4.029796</span></span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>R</span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a>          [,<span class="dv">1</span>]      [,<span class="dv">2</span>]</span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,] <span class="fl">0.2787967</span> <span class="fl">0.3878622</span></span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,] <span class="fl">0.3878622</span> <span class="fl">0.5635895</span></span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>SE.R</span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="cn">NA</span></span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>method</span>
<span id="cb25-24"><a href="#cb25-24" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Joint"</span></span>
<span id="cb25-25"><a href="#cb25-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-26"><a href="#cb25-26" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>K</span>
<span id="cb25-27"><a href="#cb25-27" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">5</span></span>
<span id="cb25-28"><a href="#cb25-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-29"><a href="#cb25-29" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>n</span>
<span id="cb25-30"><a href="#cb25-30" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">63</span></span>
<span id="cb25-31"><a href="#cb25-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-32"><a href="#cb25-32" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>iter</span>
<span id="cb25-33"><a href="#cb25-33" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="cn">NA</span></span>
<span id="cb25-34"><a href="#cb25-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-35"><a href="#cb25-35" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(,<span class="st">"class"</span>)</span>
<span id="cb25-36"><a href="#cb25-36" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"paleoTSfit"</span></span></code></pre></div>
<p><br> A multivariate random walk has a much better fit compared to
treating the traits univariately according to AICc. This indicates that
the traits are not evolving independently of each other. The estimated
<strong>R</strong> matrix indicates that the first trait has about half
the rate of evolution as the second trait and that there is substantial
covariance in the evolutionary changes of the two traits. The
evolutionary correlation can be computed by standardizing the covariance
with the product of the standard deviations on the diagonal (This can
also be using the function <code>cov3cor</code> in the
<code>stats</code> package). <br></p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fl">0.3878622</span><span class="sc">/</span>(<span class="fu">sqrt</span>(<span class="fl">0.2787967</span>)<span class="sc">*</span><span class="fu">sqrt</span>(<span class="fl">0.5635895</span>))</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">0.9784805</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Or alternatively:</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> model1<span class="ot">&lt;-</span><span class="fu">fit.multivariate.URW</span>(diam.ln_ribs.ln, <span class="at">R =</span> <span class="st">"symmetric"</span>, <span class="at">r =</span> <span class="st">"fixed"</span>)</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Model converged successfully."</span></span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> stats<span class="sc">::</span><span class="fu">cov2cor</span>(model1<span class="sc">$</span>R)</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>          [,<span class="dv">1</span>]      [,<span class="dv">2</span>]</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,] <span class="fl">1.0000000</span> <span class="fl">0.9784806</span></span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,] <span class="fl">0.9784806</span> <span class="fl">1.0000000</span></span></code></pre></div>
<p><br> A correlation of 0.98 basically means the two traits evolve as
a single trait, since at least part of the deviation from a correlation
of unity is due to measurement error.</p>
<p>Note that the <strong>R</strong> matrix is not describing the
underlying genetic or phenotypic (co)variances of the traits. The
<strong>R</strong> matrix is therefore not the same as a
<strong>P</strong> (or <strong>G</strong>) matrix in quantitative
genetics. However, the <strong>R</strong> matrix is tightly connected to
these matrices. For example, if the traits evolve only due to drift, the
<strong>R</strong> matrix is expected to be proportional to the additive
genetic variance–covariance matrix (<strong>G</strong>) (Lande 1979;
Felsenstein 1988). Estimating <strong>R</strong> can therefore aid in
evolutionary interpretations of the fossil record anchored in
evolutionary quantitative genetics.</p>
<p>Standard errors of the elements in the R matrix can be approximated
by the square roots of the diagonal elements of the inverse of the
negative of the Hessian matrix. These standard errors are automatically
estimated and reported if the argument <code>hess=TRUE</code> in
defined.</p>
<p>Parameterizing multivariate models is demanding in relation to
computational time (Felsenstein 1973; Hadfield and Nakagawa 2010;
Freckleton 2012). Be aware that fitting the multivariate models to
several traits with many samples (populations) will take much longer
time than fitting univariate models to each trait separately. One way to
somewhat follow the progress of the model fit is to set
<code>trace = TRUE</code>. This allows the user to follow the progress
of the optimization routine to minimize the likelihood function.</p>
<p>The likelihood surface of multivariate models can contain several
local peaks. It is therefore recommended to rerun the model from
different starting points (i.e., different initial parameter values).
The number of iterations can be defined by the iterations argument
(e.g., <code>iterations = 10</code>). Initial values for the search
algorithm are drawn from a normal distribution with a default standard
deviation of 1. The user can set this standard deviation by the
<code>iter.sd</code> argument. (e.g., <code>iter.sd = 0.5</code>).</p>
<p>We can check if we find evidence for a change in the
<strong>R</strong> matrix along the time-series. This involves
estimating a separate <strong>R</strong> matrix for two non-overlapping
parts of the time-series. This is done using the function
<code>fit.multivariate.URW.shift</code>. We can define the shift point
using the argument <code>shift.point</code> (e.g.,
<code>shift.point = 20</code>) or we can investigate all possible shift
points in the time series by not defining a shift point (the default
option). The length of the smallest number of samples (populations) in
each of the two segments is controlled by the <code>minb</code> argument
(the default is 10). <br></p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">fit.multivariate.URW.shift</span>(diam.ln_ribs.ln, <span class="at">hess =</span> <span class="cn">TRUE</span>)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Searching for all possible switchpoints in time series"</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>Total <span class="co"># hypotheses:  44 </span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>  [<span class="dv">1</span>] <span class="st">"Model converged successfully."</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span>  [<span class="dv">1</span>] <span class="st">"Model converged successfully."</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Omitted some of the printed output </span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a><span class="dv">44</span>  [<span class="dv">1</span>] <span class="st">"Model converged successfully."</span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>modelName</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Multivariate Random walk with two R matrices (with non-zero off-diagonal elements)"</span></span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>logL</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">189.7424</span></span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>AICc</span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">358.0887</span></span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>ancestral.values</span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">3.728940</span> <span class="fl">4.055547</span></span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-21"><a href="#cb27-21" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>R</span>
<span id="cb27-22"><a href="#cb27-22" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>R<span class="sc">$</span>R<span class="fl">.1</span></span>
<span id="cb27-23"><a href="#cb27-23" aria-hidden="true" tabindex="-1"></a>          [,<span class="dv">1</span>]      [,<span class="dv">2</span>]</span>
<span id="cb27-24"><a href="#cb27-24" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,] <span class="fl">0.3633035</span> <span class="fl">0.6331324</span></span>
<span id="cb27-25"><a href="#cb27-25" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,] <span class="fl">0.6331324</span> <span class="fl">1.1402896</span></span>
<span id="cb27-26"><a href="#cb27-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-27"><a href="#cb27-27" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>R<span class="sc">$</span>R<span class="fl">.2</span></span>
<span id="cb27-28"><a href="#cb27-28" aria-hidden="true" tabindex="-1"></a>          [,<span class="dv">1</span>]      [,<span class="dv">2</span>]</span>
<span id="cb27-29"><a href="#cb27-29" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,] <span class="fl">0.2220926</span> <span class="fl">0.2712090</span></span>
<span id="cb27-30"><a href="#cb27-30" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,] <span class="fl">0.2712090</span> <span class="fl">0.3376271</span></span>
<span id="cb27-31"><a href="#cb27-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-32"><a href="#cb27-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-33"><a href="#cb27-33" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>SE.R</span>
<span id="cb27-34"><a href="#cb27-34" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>SE.R<span class="sc">$</span>SE.R<span class="fl">.1</span></span>
<span id="cb27-35"><a href="#cb27-35" aria-hidden="true" tabindex="-1"></a>           [,<span class="dv">1</span>]       [,<span class="dv">2</span>]</span>
<span id="cb27-36"><a href="#cb27-36" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,] <span class="fl">0.01701042</span> <span class="fl">0.02975274</span></span>
<span id="cb27-37"><a href="#cb27-37" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,] <span class="fl">0.02975274</span> <span class="fl">0.06850436</span></span>
<span id="cb27-38"><a href="#cb27-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-39"><a href="#cb27-39" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>SE.R<span class="sc">$</span>SE.R<span class="fl">.2</span></span>
<span id="cb27-40"><a href="#cb27-40" aria-hidden="true" tabindex="-1"></a>            [,<span class="dv">1</span>]        [,<span class="dv">2</span>]</span>
<span id="cb27-41"><a href="#cb27-41" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,] <span class="fl">0.003426091</span> <span class="fl">0.004333208</span></span>
<span id="cb27-42"><a href="#cb27-42" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,] <span class="fl">0.004333208</span> <span class="fl">0.006481010</span></span>
<span id="cb27-43"><a href="#cb27-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-44"><a href="#cb27-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-45"><a href="#cb27-45" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>method</span>
<span id="cb27-46"><a href="#cb27-46" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Joint"</span></span>
<span id="cb27-47"><a href="#cb27-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-48"><a href="#cb27-48" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>K</span>
<span id="cb27-49"><a href="#cb27-49" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">9</span></span>
<span id="cb27-50"><a href="#cb27-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-51"><a href="#cb27-51" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>n</span>
<span id="cb27-52"><a href="#cb27-52" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">63</span></span>
<span id="cb27-53"><a href="#cb27-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-54"><a href="#cb27-54" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>iter</span>
<span id="cb27-55"><a href="#cb27-55" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="cn">NA</span></span>
<span id="cb27-56"><a href="#cb27-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-57"><a href="#cb27-57" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>parameters</span>
<span id="cb27-58"><a href="#cb27-58" aria-hidden="true" tabindex="-1"></a>shift1 </span>
<span id="cb27-59"><a href="#cb27-59" aria-hidden="true" tabindex="-1"></a>    <span class="dv">18</span> </span>
<span id="cb27-60"><a href="#cb27-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-61"><a href="#cb27-61" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>all.logl</span>
<span id="cb27-62"><a href="#cb27-62" aria-hidden="true" tabindex="-1"></a> [<span class="dv">1</span>] <span class="fl">186.3352</span> <span class="fl">184.1136</span> <span class="fl">185.2123</span> <span class="fl">188.5563</span> <span class="fl">188.9149</span> <span class="fl">186.5875</span> <span class="fl">186.0033</span> <span class="fl">189.7424</span> <span class="fl">186.3212</span> <span class="fl">186.2200</span> <span class="fl">185.5447</span></span>
<span id="cb27-63"><a href="#cb27-63" aria-hidden="true" tabindex="-1"></a>[<span class="dv">12</span>] <span class="fl">184.9384</span> <span class="fl">184.4486</span> <span class="fl">184.0647</span> <span class="fl">187.5535</span> <span class="fl">185.1279</span> <span class="fl">186.3236</span> <span class="fl">186.3285</span> <span class="fl">187.3570</span> <span class="fl">185.9061</span> <span class="fl">185.2030</span> <span class="fl">186.0518</span></span>
<span id="cb27-64"><a href="#cb27-64" aria-hidden="true" tabindex="-1"></a>[<span class="dv">23</span>] <span class="fl">186.3088</span> <span class="fl">184.8322</span> <span class="fl">184.9602</span> <span class="fl">184.2766</span> <span class="fl">185.0280</span> <span class="fl">185.3865</span> <span class="fl">184.2946</span> <span class="fl">183.7373</span> <span class="fl">182.9746</span> <span class="fl">182.5979</span> <span class="fl">181.3128</span></span>
<span id="cb27-65"><a href="#cb27-65" aria-hidden="true" tabindex="-1"></a>[<span class="dv">34</span>] <span class="fl">180.8803</span> <span class="fl">181.0370</span> <span class="fl">180.4002</span> <span class="fl">179.8530</span> <span class="fl">179.5366</span> <span class="fl">179.5296</span> <span class="fl">179.5023</span> <span class="fl">179.6067</span> <span class="fl">179.4793</span> <span class="fl">179.3429</span> <span class="fl">179.4833</span></span>
<span id="cb27-66"><a href="#cb27-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-67"><a href="#cb27-67" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>GG</span>
<span id="cb27-68"><a href="#cb27-68" aria-hidden="true" tabindex="-1"></a>     [,<span class="dv">1</span>] [,<span class="dv">2</span>] [,<span class="dv">3</span>] [,<span class="dv">4</span>] [,<span class="dv">5</span>] [,<span class="dv">6</span>] [,<span class="dv">7</span>] [,<span class="dv">8</span>] [,<span class="dv">9</span>] [,<span class="dv">10</span>] [,<span class="dv">11</span>] [,<span class="dv">12</span>] [,<span class="dv">13</span>] [,<span class="dv">14</span>] [,<span class="dv">15</span>] [,<span class="dv">16</span>] [,<span class="dv">17</span>] [,<span class="dv">18</span>]</span>
<span id="cb27-69"><a href="#cb27-69" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,]   <span class="dv">11</span>   <span class="dv">12</span>   <span class="dv">13</span>   <span class="dv">14</span>   <span class="dv">15</span>   <span class="dv">16</span>   <span class="dv">17</span>   <span class="dv">18</span>   <span class="dv">19</span>    <span class="dv">20</span>    <span class="dv">21</span>    <span class="dv">22</span>    <span class="dv">23</span>    <span class="dv">24</span>    <span class="dv">25</span>    <span class="dv">26</span>    <span class="dv">27</span>    <span class="dv">28</span></span>
<span id="cb27-70"><a href="#cb27-70" aria-hidden="true" tabindex="-1"></a>     [,<span class="dv">19</span>] [,<span class="dv">20</span>] [,<span class="dv">21</span>] [,<span class="dv">22</span>] [,<span class="dv">23</span>] [,<span class="dv">24</span>] [,<span class="dv">25</span>] [,<span class="dv">26</span>] [,<span class="dv">27</span>] [,<span class="dv">28</span>] [,<span class="dv">29</span>] [,<span class="dv">30</span>] [,<span class="dv">31</span>] [,<span class="dv">32</span>] [,<span class="dv">33</span>] [,<span class="dv">34</span>]</span>
<span id="cb27-71"><a href="#cb27-71" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,]    <span class="dv">29</span>    <span class="dv">30</span>    <span class="dv">31</span>    <span class="dv">32</span>    <span class="dv">33</span>    <span class="dv">34</span>    <span class="dv">35</span>    <span class="dv">36</span>    <span class="dv">37</span>    <span class="dv">38</span>    <span class="dv">39</span>    <span class="dv">40</span>    <span class="dv">41</span>    <span class="dv">42</span>    <span class="dv">43</span>    <span class="dv">44</span></span>
<span id="cb27-72"><a href="#cb27-72" aria-hidden="true" tabindex="-1"></a>     [,<span class="dv">35</span>] [,<span class="dv">36</span>] [,<span class="dv">37</span>] [,<span class="dv">38</span>] [,<span class="dv">39</span>] [,<span class="dv">40</span>] [,<span class="dv">41</span>] [,<span class="dv">42</span>] [,<span class="dv">43</span>] [,<span class="dv">44</span>]</span>
<span id="cb27-73"><a href="#cb27-73" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,]    <span class="dv">45</span>    <span class="dv">46</span>    <span class="dv">47</span>    <span class="dv">48</span>    <span class="dv">49</span>    <span class="dv">50</span>    <span class="dv">51</span>    <span class="dv">52</span>    <span class="dv">53</span>    <span class="dv">54</span></span>
<span id="cb27-74"><a href="#cb27-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-75"><a href="#cb27-75" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(,<span class="st">"class"</span>)</span>
<span id="cb27-76"><a href="#cb27-76" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"paleoTSfit"</span></span></code></pre></div>
<p><br> Two <strong>R</strong> matrices are returned (<em>R.1</em> and
<em>R.2</em>), along with their standard errors (since we used
<code>hess = TRUE</code>). The estimated shift point is 18
(<em>shift1</em>), which is the same shift point we detected when
fitting the Bokma model to the diameter time series. The rate of
evolution is much larger in the second trait compared to the first trait
(the diagonal elements in <em>R.1</em>) while the difference in rate of
evolution is much smaller after the shift point (the diagonal elements
in <em>R.2</em>). The evolutionary correlation is 0.98 and 0.99 before
and after the shift point. The number of parameters (<em>K</em>) for
this model is 9 compared to 5 for the model where we estimated a single
<strong>R</strong> matrix for the multivariate data set. The model with
two <strong>R</strong> matrices has a lower (better) AICc score compared
to the model with a single <strong>R</strong> matrix. This difference in
AICc scores is likely to be a result of the differences in rates of
evolution in <em>R.1</em> and <em>R.2</em> since the evolutionary
correlation is estimated to be very similar in both <em>R.1</em> and
<em>R.2</em>.</p>
<p>We now check if the multivariate accelerated and decelerated models
have a better fit than the multivariate random walk models. <br></p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> multi.accel<span class="ot">&lt;-</span><span class="fu">fit.multivariate.URW</span>(diam.ln_ribs.ln, <span class="at">R =</span> <span class="st">"symmetric"</span>, <span class="at">r =</span> <span class="st">"accel"</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Model converged successfully."</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> multi.decel<span class="ot">&lt;-</span><span class="fu">fit.multivariate.URW</span>(diam.ln_ribs.ln, <span class="at">R =</span> <span class="st">"symmetric"</span>, <span class="at">r =</span> <span class="st">"decel"</span>)</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Model converged successfully."</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> multi.accel<span class="sc">$</span>AICc;multi.decel<span class="sc">$</span>AICc</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">336.454</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">351.5167</span></span></code></pre></div>
<p><br> The multivariate decelerated evolution model has a similar
(but slightly worse) fit to the data compared to the multivariate
unbiased random walk model with a single <strong>R</strong> matrix, but
is out competed by the model with a mode shift estimating two
<strong>R</strong> matrices.</p>
</div>
<div class="section level3">
<h3 id="multivariate-ornstein-uhlenbeck-models">5.2 Multivariate Ornstein-Uhlenbeck models<a class="anchor" aria-label="anchor" href="#multivariate-ornstein-uhlenbeck-models"></a>
</h3>
<p>Multivariate Ornstein-Uhlenbeck models allow for testing a range of
different adaptive hypotheses of trait evolution, as detailed below.
Adapted to describe evolution of traits within the same lineage, the
multivariate Ornstein-Uhlenbeck process is described by the following
differential equation (Bartoszek <em>et al.</em> 2012; Reitan <em>et
al.</em> 2012; Clavel <em>et al.</em> 2015): <br><br><span class="math display">\[dY = A(\theta(t)-Y(t))dt + RdW(t)\]</span> <br>
where <strong>A</strong> is square matrix with dimensions equal to the
number of investigated traits that describes the rate of evolution
toward the optimal trait values (<strong>A</strong> is often called the
pull matrix), <span class="math inline">\(\theta\)</span> is a vector
containing the optimum for each trait, <strong>R</strong> is a square
matrix with dimensions equal to the number of investigated traits
describing the stochastic changes in the traits (<strong>R</strong> is
often called the drift matrix), and <span class="math inline">\(W\)</span> is the diffusion parameter. Under the
assumption that the we only have one selective regime (optimum) per
trait, the expected trait mean of the Ornstein-Uhlenbeck process is the
weighted sum of the optimum and the root value (Hansen 1997): <br><br><span class="math display">\[E\left[Z_{i}\right]=
e^{(-At_{i})}z_{0} + (1-e^{(-At_{i})}\theta)\]</span> <br> where <span class="math inline">\(Z_{i}\)</span> is the expected trait value for
sample <em>i</em>, <span class="math inline">\(z_{0}\)</span> is the
ancestral trait mean, <em>Θ</em> is the optimum, <strong>A</strong> is
the pull matrix measuring the rate of adaptation to the optimum, and
<span class="math inline">\(t_{i}\)</span> is the time interval from the
ancestral population mean (the start of the time-series, which has a
time of 0) to the <em>i</em>th population mean.</p>
<p>The variance and covariance of sample/population means are given by
the following expression: <br><br><span class="math display">\[Cov(z_{i}, z_{j})=\left[P\left( \left[
\frac{1}{\lambda_{k} + \lambda_{l}} \left( 1-e^{-({\lambda_{k} +
\lambda_{l})}t_{a}}\right)\right]_{1\le kl \le m}   \odot
P^{-1}\Sigma\Sigma^{T}(P^{-1})^{T}\right)
P^{T}\right]e^{-A^{T}t_{ij}}\]</span> <br> where <strong>P</strong> is
the orthogonal matrix of eigenvectors of <strong>A</strong>, <span class="math inline">\(\Sigma\Sigma^{T}\)</span> is the Cholesky
decomposition of the <strong>R</strong> matrix, <span class="math inline">\(\lambda_{i}\)</span> is the <em>i</em>th
eigenvalue of <strong>A</strong>, <span class="math inline">\(t_{a}\)</span> is the time interval from the
ancestral population to the oldest of the two populations <span class="math inline">\(z_{i}\)</span> and <span class="math inline">\(z_{j}\)</span>, and <span class="math inline">\(t_{ij}\)</span> is the time separating two samples
<span class="math inline">\(z_{i}\)</span> and <span class="math inline">\(z_{j}\)</span>, while <span class="math inline">\(\odot\)</span> represents the Hadamard
(element-wise) product.</p>
<p>Different ways to parameterize the <strong>A</strong> matrix allows
testing a range of evolutionary hypotheses using the multivariate OU
process. While the <strong>R</strong> matrix in the multivariate version
of the unbiased random walk (and in the multivariate OU model) either
needs to be diagonal (only elements on the diagonal while off-diagonal
elements are zero) or complete (in the sense that both the diagonal and
off-diagonal elements are non-zero), this is not the case for the
<strong>A</strong> matrix. Generally, we can divide up the hypotheses
being tested using the multivariate OU into four different types:
<br></p>
<ul>
<li><p>Independent evolution (only diagonal elements in both the
<strong>A</strong> and the <strong>R</strong> matrix, i.e. equivalent as
fitting univariate models to each trait separately.)</p></li>
<li><p>Independent adaptation (only diagonal elements in the
<strong>A</strong> matrix while the <strong>R</strong> matrix is
completely parameterized. In such a model, the traits are adapting
independently toward their optima, but the stochastic changes in the
traits are correlated.)</p></li>
<li><p>At least one trait affects the optimum of the other trait. The
diagonal and at least one of the off-diagonal elements in the
<strong>A</strong> matrix is non-zero. It is the trait in the column of
the non-zero off-diagonal element that affects the optimum of the trait
in the row of the non-zero off-diagonal element. A negative off-diagonal
element means the trait evolves toward the optimum determined by the the
other trait, while a positive off-diagonal element means the trait
evolve away from the optimum determined by the other trait. The the
stochastic changes in the trait can be either correlated or
non-correlated.</p></li>
<li><p>The same traits affect each others optimum (but to different
degrees). The same off-diagonal elements on both sides of the diagonal
are parameterized in the <strong>A</strong> matrix. The the stochastic
changes in the trait can be either correlated or non-correlated. <br>
An <strong>A</strong> matrix with off-diagonal elements investigates
<a href="https://en.wikipedia.org/wiki/Granger_causality" class="external-link">(Granger)</a>
causality between the two traits/variables (Granger 1969; Schweder
1970). Simply speaking, we have evidence for Granger causality if
observations in one time series is useful for forecasting observations
in one or several other time series, which would be the case if one
trait affects the optimum that another trait tracks with a lag.
Multivariate OU models therefore enable us to move beyond interpreting
correlations among variables. While correlation is a measure of linear
dependence between two random variables, the time dimension (i.e. the
lag/evolutionary inertia in the tracking of the changing optimum) of the
potential relationship between two variables is important in Granger
causality. And while a correlation is symmetric (the correlation between
X and Y is the same as the correlation between Y and X), this is not
necessarily the case for Granger causality. X can (Granger) cause Y,
without Y Granger-causing X.</p></li>
</ul>
<p>For a data set consisting of two traits, there are four possible
parameterizations of the <strong>A</strong> matrix (see panels below).
Independent adaptation of both traits (panel a), trait Y affecting the
optimum of trait X (panel b), trait X affecting the optimum of trait Y,
and the traits X and Y affecting each others optima (panel d). Each of
these four ways of parameterizing the <strong>A</strong> matrix can be
combined with an <strong>R</strong> matrix with elements only on the
diagonal (the stochastic changes in the traits are uncorrelated) or a
completely parameterized <strong>R</strong> matrix (the stochastic
changes in the traits are correlated).<br><br><img src="A_matrices.png" width="40%" style="display: block; margin: auto;"><br> It is also possible to implement a model where the last trait in
the data set evolves as an unbiased random walk which affects the optima
for all other traits in the data set. This model can be fitted by
defining <code>A.matrix ="OUBM"</code>, which sets the last diagonal
element in the <strong>A</strong> matrix to zero. A zero diagonal
element in the <strong>A</strong> matrix means there is no tendency for
this trait to evolve towards an optimal value (which is the case for an
unbiased random walk). Setting <code>A.matrix ="OUBM"</code> will
automatically define the <strong>R</strong> matrix to only have diagonal
elements (i.e. this option for the <strong>A</strong> matrix will
override how the <strong>R</strong> matrix is defined by the user).</p>
<p>Multivariate Ornstein-Uhlenbeck models can be fitted using two
different functions in <code>evoTS</code>. The function
<code>fit.multivariate.OU</code> allows the user to use pre-defined
arguments for the <strong>A</strong> and <strong>R</strong> matrices to
parameterize these matrices. The <code>A.matrix</code> can either be
defined as “diag” (panel a above), “upper.tri” (panel b above),
“lower.tri” (panel c above), and “full” (panel d above). Default is
“diag”. The <code>R.matrix</code> can be defined as “diag” or
“symmetric”.</p>
<p>The function <code>fit.multivariate.OU.user.defined</code> lets the
user define which elements in the <strong>A</strong> and
<strong>R</strong> matrices that will be parameterized. This function
allows full flexibility for the type of hypotheses that are being tested
as the parameterization is not constrained to follow the pre-defined
options for the <strong>A</strong> matrix available in the function
<code>fit.multivariate.OU</code>.</p>
<p>We will investigate whether the two traits from the <em>S.
yellowstonensis</em> lineage show evidence of independent evolution
(only diagonal elements in the <strong>A</strong> and <strong>R</strong>
matrices) or whether only the adaptation part of the trait dynamics is
independent in the two traits (diagonal <strong>A</strong> matrix and
symmetric <strong>R</strong> matrix). We will test these hypotheses
using the <code>fit.multivariate.OU</code> function. <br></p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> OUOU.model1<span class="ot">&lt;-</span><span class="fu">fit.multivariate.OU</span>(diam.ln_ribs.ln, <span class="at">A.matrix=</span><span class="st">"diag"</span>, <span class="at">R.matrix=</span><span class="st">"diag"</span>)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> OUOU.model2<span class="ot">&lt;-</span><span class="fu">fit.multivariate.OU</span>(diam.ln_ribs.ln, <span class="at">A.matrix=</span><span class="st">"diag"</span>, <span class="at">R.matrix=</span><span class="st">"symmetric"</span>)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> OUOU.model1<span class="sc">$</span>AICc;OUOU.model2<span class="sc">$</span>AICc</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">267.3968</span></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">352.6937</span></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> OUOU.model2</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>logL</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">187.045</span></span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>AICc</span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">352.6937</span></span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>ancestral.values</span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">3.723682</span> <span class="fl">4.019902</span></span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>optima</span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">3.884589</span> <span class="fl">4.292722</span></span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-19"><a href="#cb29-19" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>A</span>
<span id="cb29-20"><a href="#cb29-20" aria-hidden="true" tabindex="-1"></a>         [,<span class="dv">1</span>]     [,<span class="dv">2</span>]</span>
<span id="cb29-21"><a href="#cb29-21" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,] <span class="fl">11.32464</span>  <span class="fl">0.00000</span></span>
<span id="cb29-22"><a href="#cb29-22" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,]  <span class="fl">0.00000</span> <span class="fl">14.16996</span></span>
<span id="cb29-23"><a href="#cb29-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-24"><a href="#cb29-24" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>half.life</span>
<span id="cb29-25"><a href="#cb29-25" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">0.06120697</span> <span class="fl">0.04891666</span></span>
<span id="cb29-26"><a href="#cb29-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-27"><a href="#cb29-27" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>R</span>
<span id="cb29-28"><a href="#cb29-28" aria-hidden="true" tabindex="-1"></a>           [,<span class="dv">1</span>]       [,<span class="dv">2</span>]</span>
<span id="cb29-29"><a href="#cb29-29" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,] <span class="fl">0.01985206</span> <span class="fl">0.07543067</span></span>
<span id="cb29-30"><a href="#cb29-30" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,] <span class="fl">0.07543067</span> <span class="fl">0.91041225</span></span>
<span id="cb29-31"><a href="#cb29-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-32"><a href="#cb29-32" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>method</span>
<span id="cb29-33"><a href="#cb29-33" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Joint"</span></span>
<span id="cb29-34"><a href="#cb29-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-35"><a href="#cb29-35" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>K</span>
<span id="cb29-36"><a href="#cb29-36" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">9</span></span>
<span id="cb29-37"><a href="#cb29-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-38"><a href="#cb29-38" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>n</span>
<span id="cb29-39"><a href="#cb29-39" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">63</span></span>
<span id="cb29-40"><a href="#cb29-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-41"><a href="#cb29-41" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>iter</span>
<span id="cb29-42"><a href="#cb29-42" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="cn">NA</span></span>
<span id="cb29-43"><a href="#cb29-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-44"><a href="#cb29-44" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(,<span class="st">"class"</span>)</span>
<span id="cb29-45"><a href="#cb29-45" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"paleoTSfit"</span></span>
<span id="cb29-46"><a href="#cb29-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-47"><a href="#cb29-47" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span>stats<span class="sc">::</span><span class="fu">cov2cor</span>(OUOU.model2<span class="sc">$</span>R)</span>
<span id="cb29-48"><a href="#cb29-48" aria-hidden="true" tabindex="-1"></a>          [,<span class="dv">1</span>]      [,<span class="dv">2</span>]</span>
<span id="cb29-49"><a href="#cb29-49" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,] <span class="fl">1.0000000</span> <span class="fl">0.5612106</span></span>
<span id="cb29-50"><a href="#cb29-50" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,] <span class="fl">0.5612106</span> <span class="fl">1.0000000</span></span></code></pre></div>
<p><br> The model with independent adaptation and correlated
stochastic changes is much better than the independent evolution model.
The half-life for the log diameter and log number of ribs are 6.1% and
4.9% of the length of the time-series, which translates into (13728 *
0.061 =) 837 and (13728 * 0.049 =) 673 years respectively. The
correlation of the stochastic changes is substantial, but much reduced
compared to the estimate of the correlation from the multivariate
unbiased random walk. This is because a substantial part of the trait
dynamics in a multivariate OU model is due to the deterministic approach
of the traits towards the optima. The model has an almost identical
relative fit compared to the multivariate unbiased random walk according
to AICc, but is out competed by the unbiased random walk with a a mode
shift.</p>
<p>We now test if a more complex parameterization of the
<strong>A</strong> matrix gives us a better relative model fit according
to AICc. <br></p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> OUOU.model3<span class="ot">&lt;-</span><span class="fu">fit.multivariate.OU</span>(diam.ln_ribs.ln, <span class="at">A.matrix=</span><span class="st">"upper.tri"</span>, <span class="at">R.matrix=</span><span class="st">"symmetric"</span>)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> OUOU.model4<span class="ot">&lt;-</span><span class="fu">fit.multivariate.OU</span>(diam.ln_ribs.ln, <span class="at">A.matrix=</span><span class="st">"full"</span>, <span class="at">R.matrix=</span><span class="st">"symmetric"</span>)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> OUOU.model5<span class="ot">&lt;-</span><span class="fu">fit.multivariate.OU</span>(diam.ln_ribs.ln, <span class="at">A.matrix=</span><span class="st">"OUBM"</span>)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> OUOU.model3<span class="sc">$</span>AICc;OUOU.model4<span class="sc">$</span>AICc;OUOU.model5<span class="sc">$</span>AICc</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">349.9107</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">349.8412</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">301.516</span></span></code></pre></div>
<p><br> None of the models with off-diagonal elements in the
<strong>A</strong> matrix show a better relative fit than the model with
a diagonal <strong>A</strong> and a symmetric <strong>R</strong> matrix.
The best model we have investigated remains the multivariate unbiased
random walk with a mode shift, allowing two different unbiased random
walks to explain the trait dynamics.</p>
<p>However, before we trust this result, we should make sure to run the
multivariate models from different initial starting values, this to
increase our chances of finding a potentially higher peak on the
log-likelihood surface. The number of iterations can be defined by the
<code>iterations</code> argument. The starting values are drawn from a
normal distribution with a standard deviation of one. The user can
define a larger or smaller standard deviation using the argument
<code>iter.sd</code>. <code>trace = TRUE</code> allows the user to keep
an eye on how the optimization proceeds.</p>
<p>To illustrate how the user can define which elements in the
<strong>A</strong> and <strong>R</strong> matrices that should be
parameterized using the function
<code>fit.multivariate.OU.user.defined</code>, we can fit a lower
triangle <strong>A</strong> matrix and a symmetric <strong>R</strong>
matrix. A lower triangle <strong>A</strong> matrix means the diameter is
affecting the optimum of the number of ribs. Which elements in
<strong>A</strong> and <strong>R</strong> that should be parameterized
or set to zero is given by 1 and 0 respectively. <br></p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> A <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">1</span>), <span class="at">nrow=</span><span class="dv">2</span>, <span class="at">byrow=</span><span class="cn">TRUE</span>)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> R <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>), <span class="at">nrow=</span><span class="dv">2</span>, <span class="at">byrow=</span><span class="cn">TRUE</span>)</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> OUOU.model6<span class="ot">&lt;-</span><span class="fu">fit.multivariate.OU.user.defined</span>(diam.ln_ribs.ln, <span class="at">A.user=</span>A, <span class="at">R.user=</span>R)</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> OUOU.model6<span class="sc">$</span>AICc</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">349.357</span></span></code></pre></div>
<p><br><br></p>
</div>
</div>
<div class="section level2">
<h2 id="references">6.0 References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Akaike. H. 1974. A new look at the statistical model identification.
<em>IEEE Transactions on Automatic Control</em> 19:716 - 723</p>
<p>Bartoszek, K., J. Pienaar, P. Mostad, S. Andersson, and T. F. Hansen.
2012 A phylogenetic comparative method for studying multivariate
adaptation. <em>Journal of Theoretical Biology</em> 314:204–215.</p>
<p>Bokma, F. 2002. Detection of punctuated equilibrium from molecular
phylogenies. <em>Journal of Evolutionary Biology</em> 15:1048–1056.</p>
<p>Burnham, K. P. and D. R. Anderson. 2002. Model Selection and
multimodel inference: a practical information-theoretic approach (2nd
ed.)- Springer-Verlag.</p>
<p>Clavel, J., G. Escarguel, and G. Merceron. 2015. mvmorph: an r
package for fitting multivariate evolutionary models to morphometric
data. <em>Methods in Ecology and Evolution</em> 6:1311–1319.</p>
<p>Cooper, N., and A. Purvis, 2010. Body size evolution in mammals:
complexity in tempo and mode. <em>The American Naturalist</em>
175:727–738.</p>
<p>Edwards, A. W. F. 1992. Likelihood. expanded edition Johns Hopkins
University Press. Baltimore, MD.</p>
<p>Felsenstein, J. 1973. Maximum-likelihood estimation of evolutionary
trees from continuous characters. <em>The American Journal of Human
Genetics</em> 25:471–492.</p>
<p>Felsenstein, J. 1988. Phylogenies and quantitative characters.
<em>Annual Review of Ecology, Evolution, and Systematics</em>
19:445–471.</p>
<p>Freckleton R. P. 2012. Fast likelihood calculations for comparative
analyses. <em>Methods in Ecology and Evolution</em> 3:940-947.</p>
<p>Granger, C. W. J. 1969. Investigating causal relations by econometric
models and cross‐spectral methods. <em>Econometrica</em> 37:424-438.</p>
<p>Hadfield, J. D., and S. Nakagawa. 2010. General quantitative genetic
methods for comparative biology: phylogenies, taxonomies and multi-trait
models for continuous and categorical characters. <em>Journal of
Evolutionary Biology</em> 23:494-508.</p>
<p>Hansen, T. F. 1997. Stabilizing selection and the comparative
analysis of adaptation. <em>Evolution</em> 51:1341–1351.</p>
<p>Hansen, T. F., J. Pienaar, and S. H. Orzack. 2008. A comparative
method for studying adaptation to a randomly evolving environment.
<em>Evolution</em> 62:1965–1977.</p>
<p>Harmon, L. J. et al. 2010. Early bursts of body size and shape
evolution are rare in comparative data. <em>Evolution</em>
64:2385–2396.</p>
<p>Hunt, G. 2006. Fitting and comparing models of phyletic evolution:
random walks and beyond. <em>Paleobiology</em> 32:578–601.</p>
<p>Hunt, G. 2008a. Gradual or pulsed evolution: when should
punctuational explanations be preferred? <em>Paleobiology</em>
34:360–377.</p>
<p>Hunt, G. 2008b. Evolutionary patterns within fossil lineages:
model-based assessment of modes, rates, punctuations and process. In
R.K. Bambach and P.H. Kelley, eds. From Evolution to Geobiology:
Research Questions Driving Paleontology at the Start of a New
Century:578–601.</p>
<p>Hunt, G., M. Bell &amp; M. Travis. 2008. Evolution towards a new
adaptive optimum: phenotypic evolution in a fossil stickleback lineage.
<em>Evolution</em> 62:700-710.</p>
<p>Hunt, G., S. Wicaksono, J. E. Brown, and G. K. Macleod. 2010.
Climate-driven body size trends in the ostracod fauna of the deep Indian
Ocean. <em>Palaeontology</em> 53:1255-1268.</p>
<p>Hunt, G. 2013. Testing the link between phenotypic evolution and
speciation: an integrated palaeontological and phylogenetic analysis.
<em>Methods in Ecololgy and Evolution</em> 4:714–723.</p>
<p>Hunt, G., M. J. Hopkins, and S. Lidgard. 2015 Simple versus complex
models of trait evolution and stasis as a response to environmental
change. <em>PNAS</em> 112:4885–4890.</p>
<p>Ingram, T. 2011. Speciation along a depth gradient in a marine
adaptive radiation. Proceedings og the Royal Society B Biological
Sciences 278:613–618.</p>
<p>Lande, R. 1979. Quantitative genetic analysis of multivariate
evolution, applied to brain: body size allometry. <em>Evolution</em>
33:402-416.</p>
<p>Reitan, T., T. Schweder, and J. Henderiks. 2012. Phenotypic evolution
studied by layered stochastic differential equations. <em>The Annals of
Applied Statistics</em>. 6:1531–1551.</p>
<p>Revell, L. J., and L. Harmon 2008. Testing quantitative genetic
hypotheses about the evolutionary rate matrix for continuous characters.
<em>Evolutionary Ecology Research</em> 10:311–331.</p>
<p>Schweder, T. 1970. Composable Markov processes. <em>Journal of
Applied Probability</em> 7:400–410.</p>
<p>Theriot, E. C., S. C., Fritz, C. Whitlock, C. and D. J. Conley. 2006.
Late Quaternary rapid morphological evolution of an endemic diatom in
Yellowstone Lake, Wyoming. <em>Paleobiology</em> 32:38–54.</p>
<p>Voje, K. L. 2020. Testing eco‐evolutionary predictions using fossil
data: Phyletic evolution following ecological opportunity.
<em>Evolution</em> 74:188–200.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Kjetil Lysne Voje.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
